// @deriverse/kit/dist/auto_data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoData = void 0;
const kit_1 = require("@solana/kit");
class AutoData {
    get length() {
        return this.buffer.length;
    }
    get offset() {
        return this.offset_;
    }
    constructor(data, offset = 0) {
        this.buffer = Buffer.from((0, kit_1.getBase64Encoder)().encode(data[0]));
        this.offset_ = offset;
    }
    readU8() {
        const value = this.buffer.readUint8(this.offset_);
        this.offset_ += 1;
        return value;
    }
    readU16() {
        const value = this.buffer.readUint16LE(this.offset_);
        this.offset_ += 2;
        return value;
    }
    readU32() {
        const value = this.buffer.readUint32LE(this.offset_);
        this.offset_ += 4;
        return value;
    }
    readI64() {
        const value = Number(this.buffer.readBigInt64LE(this.offset_));
        this.offset_ += 8;
        return value;
    }
    readF64() {
        const value = this.buffer.readDoubleLE(this.offset_);
        this.offset_ += 8;
        return value;
    }
    readAddress() {
        const value = (0, kit_1.getAddressDecoder)().decode(this.buffer.slice(this.offset_, this.offset_ + 32));
        this.offset_ += 32;
        return value;
    }
}
exports.AutoData = AutoData;
