// @deriverse/kit/dist/auto_buffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoBuffer = void 0;
class AutoBuffer {
    get length() {
        return this.buffer.length;
    }
    get offset() {
        return this.offset_;
    }
    constructor(buffer, offset = 0) {
        this.buffer = buffer;
        this.offset_ = offset;
    }
    readU8() {
        const value = this.buffer.readUint8(this.offset_);
        this.offset_ += 1;
        return value;
    }
    readU16() {
        const value = this.buffer.readUint16LE(this.offset_);
        this.offset_ += 2;
        return value;
    }
    readU32() {
        const value = this.buffer.readUint32LE(this.offset_);
        this.offset_ += 4;
        return value;
    }
    readI64() {
        const value = Number(this.buffer.readBigInt64LE(this.offset_));
        this.offset_ += 8;
        return value;
    }
    readF64() {
        const value = this.buffer.readDoubleLE(this.offset_);
        this.offset_ += 8;
        return value;
    }
}
exports.AutoBuffer = AutoBuffer;
