// @deriverse/kit/dist/instruction_models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activateClientRefProgramData = exports.garbageCollectorData = exports.changeVotingData = exports.setSeatPurchasingFeeData = exports.perpClientsProcessingData = exports.newBaseCrncyData = exports.changeDenominatorData = exports.setVarianceData = exports.pointsProgramExpiration = exports.newPrivateClient = exports.sellMarketSeatData = exports.buyMarketSeatData = exports.changeRefProgramData = exports.perpMassCancelData = exports.perpOrderCancelData = exports.setInstrReadyForPerpUpgradeData = exports.upgradeToPerpData = exports.airdropData = exports.votingData = exports.perpQuotesReplaceData = exports.spotQuotesReplaceData = exports.swapData = exports.withdrawData = exports.perpWithdrawData = exports.moveSpotAvailFundsData = exports.perpDepositData = exports.feesWithdrawData = exports.feesDepositData = exports.depositData = exports.newInstrumentData = exports.spotLpData = exports.spotMassCancelData = exports.spotOrderCancelData = exports.perpStatisticsResetData = exports.perpChangeLeverageData = exports.newPerpOrderData = exports.newSpotOrderData = exports.newRootAccountData = exports.newOperatorData = void 0;
function newOperatorData(tag) {
    let buf = Buffer.alloc(4);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    return buf;
}
exports.newOperatorData = newOperatorData;
function newRootAccountData(tag, privateMode, lutSlot) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(privateMode, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(lutSlot, 4);
    return buf;
}
exports.newRootAccountData = newRootAccountData;
function newSpotOrderData(tag, ioc, orderType, side, instrId, price, amount, edgePrice) {
    let buf = Buffer.alloc(32);
    buf.writeUint8(tag, 0);
    buf.writeUint8(ioc, 1);
    buf.writeUint8(orderType, 2);
    buf.writeUint8(side, 3);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(price)), 8);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 16);
    buf.writeBigInt64LE(BigInt(Math.floor(edgePrice)), 24);
    return buf;
}
exports.newSpotOrderData = newSpotOrderData;
function newPerpOrderData(tag, ioc, leverage, orderType, side, instrId, price, amount, edgePrice) {
    let buf = Buffer.alloc(40);
    buf.writeUint8(tag, 0);
    buf.writeUint8(ioc, 1);
    buf.writeUint8(leverage, 2);
    buf.writeUint8(orderType, 3);
    buf.writeUint8(side, 4);
    buf.writeUint8(0, 5);
    buf.writeUint16LE(0, 6);
    buf.writeUint32LE(0, 8);
    buf.writeUint32LE(instrId, 12);
    buf.writeBigInt64LE(BigInt(Math.floor(price)), 16);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 24);
    buf.writeBigInt64LE(BigInt(Math.floor(edgePrice)), 32);
    return buf;
}
exports.newPerpOrderData = newPerpOrderData;
function perpChangeLeverageData(tag, leverage, instrId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(leverage, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.perpChangeLeverageData = perpChangeLeverageData;
function perpStatisticsResetData(tag, instrId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.perpStatisticsResetData = perpStatisticsResetData;
function spotOrderCancelData(tag, side, instrId, orderId) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(side, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(orderId)), 8);
    return buf;
}
exports.spotOrderCancelData = spotOrderCancelData;
function spotMassCancelData(tag, instrId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.spotMassCancelData = spotMassCancelData;
function spotLpData(tag, side, instrId, amount, minPrice, maxPrice) {
    let buf = Buffer.alloc(32);
    buf.writeUint8(tag, 0);
    buf.writeUint8(side, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 8);
    buf.writeBigInt64LE(BigInt(Math.floor(minPrice)), 16);
    buf.writeBigInt64LE(BigInt(Math.floor(maxPrice)), 24);
    return buf;
}
exports.spotLpData = spotLpData;
function newInstrumentData(tag, crncyTokenId, lutSlot, price) {
    let buf = Buffer.alloc(24);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(0, 4);
    buf.writeUint32LE(crncyTokenId, 8);
    buf.writeUint32LE(lutSlot, 12);
    buf.writeBigInt64LE(BigInt(Math.floor(price)), 16);
    return buf;
}
exports.newInstrumentData = newInstrumentData;
function depositData(tag, competitionId, depositAll, tokenId, amount, lutSlot, refId) {
    let buf = Buffer.alloc(24);
    buf.writeUint8(tag, 0);
    buf.writeUint8(competitionId, 1);
    buf.writeUint8(depositAll, 2);
    buf.writeUint8(0, 3);
    buf.writeUint32LE(tokenId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 8);
    buf.writeUint32LE(lutSlot, 16);
    buf.writeUint32LE(refId, 20);
    return buf;
}
exports.depositData = depositData;
function feesDepositData(tag, tokenId, amount) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(tokenId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 8);
    return buf;
}
exports.feesDepositData = feesDepositData;
function feesWithdrawData(tag, tokenId, amount) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(tokenId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 8);
    return buf;
}
exports.feesWithdrawData = feesWithdrawData;
function perpDepositData(tag, instrId, amount) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 8);
    return buf;
}
exports.perpDepositData = perpDepositData;
function moveSpotAvailFundsData(tag, instrId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.moveSpotAvailFundsData = moveSpotAvailFundsData;
function perpWithdrawData(tag, instrId, amount) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 8);
    return buf;
}
exports.perpWithdrawData = perpWithdrawData;
function withdrawData(tag, tokenId, amount) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(tokenId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 8);
    return buf;
}
exports.withdrawData = withdrawData;
function swapData(tag, inputCrncy, instrId, price, amount) {
    let buf = Buffer.alloc(24);
    buf.writeUint8(tag, 0);
    buf.writeUint8(inputCrncy, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(price)), 8);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 16);
    return buf;
}
exports.swapData = swapData;
function spotQuotesReplaceData(tag, instrId, newBidPrice, newBidQty, oldBidOrderId, newAskPrice, newAskQty, oldAskOrderId) {
    let buf = Buffer.alloc(56);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(newBidPrice)), 8);
    buf.writeBigInt64LE(BigInt(Math.floor(newBidQty)), 16);
    buf.writeBigInt64LE(BigInt(Math.floor(oldBidOrderId)), 24);
    buf.writeBigInt64LE(BigInt(Math.floor(newAskPrice)), 32);
    buf.writeBigInt64LE(BigInt(Math.floor(newAskQty)), 40);
    buf.writeBigInt64LE(BigInt(Math.floor(oldAskOrderId)), 48);
    return buf;
}
exports.spotQuotesReplaceData = spotQuotesReplaceData;
function perpQuotesReplaceData(tag, instrId, newBidPrice, newBidQty, oldBidOrderId, newAskPrice, newAskQty, oldAskOrderId) {
    let buf = Buffer.alloc(56);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(newBidPrice)), 8);
    buf.writeBigInt64LE(BigInt(Math.floor(newBidQty)), 16);
    buf.writeBigInt64LE(BigInt(Math.floor(oldBidOrderId)), 24);
    buf.writeBigInt64LE(BigInt(Math.floor(newAskPrice)), 32);
    buf.writeBigInt64LE(BigInt(Math.floor(newAskQty)), 40);
    buf.writeBigInt64LE(BigInt(Math.floor(oldAskOrderId)), 48);
    return buf;
}
exports.perpQuotesReplaceData = perpQuotesReplaceData;
function votingData(tag, choice, votingCounter) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(choice, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(votingCounter, 4);
    return buf;
}
exports.votingData = votingData;
function airdropData(tag, ratio) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(0, 4);
    return buf;
}
exports.airdropData = airdropData;
function upgradeToPerpData(tag, instrId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.upgradeToPerpData = upgradeToPerpData;
function setInstrReadyForPerpUpgradeData(tag, instrId, variance) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.setInstrReadyForPerpUpgradeData = setInstrReadyForPerpUpgradeData;
function perpOrderCancelData(tag, side, instrId, orderId) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(side, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(orderId)), 8);
    return buf;
}
exports.perpOrderCancelData = perpOrderCancelData;
function perpMassCancelData(tag, instrId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.perpMassCancelData = perpMassCancelData;
function changeRefProgramData(tag, refProgramDuration, refLinkDuration, refDiscount, refRatio) {
    let buf = Buffer.alloc(32);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(0, 4);
    buf.writeUint32LE(refProgramDuration, 8);
    buf.writeUint32LE(refLinkDuration, 12);
    return buf;
}
exports.changeRefProgramData = changeRefProgramData;
function buyMarketSeatData(tag, instrId, edgePrice, amount) {
    let buf = Buffer.alloc(24);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(edgePrice)), 8);
    buf.writeBigInt64LE(BigInt(Math.floor(amount)), 16);
    return buf;
}
exports.buyMarketSeatData = buyMarketSeatData;
function sellMarketSeatData(tag, instrId, edgePrice) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    buf.writeBigInt64LE(BigInt(Math.floor(edgePrice)), 8);
    return buf;
}
exports.sellMarketSeatData = sellMarketSeatData;
function newPrivateClient(tag, expirationTime) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(expirationTime, 4);
    return buf;
}
exports.newPrivateClient = newPrivateClient;
function pointsProgramExpiration(tag, newExpirationTime) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(newExpirationTime, 4);
    return buf;
}
exports.pointsProgramExpiration = pointsProgramExpiration;
function setVarianceData(tag, variance) {
    let buf = Buffer.alloc(12);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    return buf;
}
exports.setVarianceData = setVarianceData;
function changeDenominatorData(tag, baseCrncyId, denominator) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(baseCrncyId, 4);
    return buf;
}
exports.changeDenominatorData = changeDenominatorData;
function newBaseCrncyData(tag, denominator) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(0, 4);
    return buf;
}
exports.newBaseCrncyData = newBaseCrncyData;
function perpClientsProcessingData(tag, instrId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.perpClientsProcessingData = perpClientsProcessingData;
function setSeatPurchasingFeeData(tag, fee) {
    let buf = Buffer.alloc(16);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(0, 4);
    return buf;
}
exports.setSeatPurchasingFeeData = setSeatPurchasingFeeData;
function changeVotingData(tag, newChoice, votingCounter) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(newChoice, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(votingCounter, 4);
    return buf;
}
exports.changeVotingData = changeVotingData;
function garbageCollectorData(tag, instrId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(instrId, 4);
    return buf;
}
exports.garbageCollectorData = garbageCollectorData;
function activateClientRefProgramData(tag, refId) {
    let buf = Buffer.alloc(8);
    buf.writeUint8(tag, 0);
    buf.writeUint8(0, 1);
    buf.writeUint16LE(0, 2);
    buf.writeUint32LE(refId, 4);
    return buf;
}
exports.activateClientRefProgramData = activateClientRefProgramData;
