// @deriverse/kit/dist/logs_models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MoveSpotAvailFundsReportModel = exports.SpotOrderRevokeReportModel = exports.PerpOrderRevokeReportModel = exports.SpotOrderCancelReportModel = exports.PerpOrderCancelReportModel = exports.SpotNewOrderReportModel = exports.PerpNewOrderReportModel = exports.PerpSocLossReportModel = exports.PerpFundingReportModel = exports.SpotFeesReportModel = exports.PerpFeesReportModel = exports.SpotMassCancelReportModel = exports.PerpMassCancelReportModel = exports.SpotPlaceMassCancelReportModel = exports.PerpPlaceMassCancelReportModel = exports.SwapOrderReportModel = exports.SpotPlaceOrderReportModel = exports.PerpPlaceOrderReportModel = exports.SpotFillOrderReportModel = exports.PerpFillOrderReportModel = exports.SpotlpTradeReportModel = exports.PerpWithdrawReportModel = exports.WithdrawReportModel = exports.SellMarketSeatReportModel = exports.BuyMarketSeatReportModel = exports.PerpDepositReportModel = exports.FeesWithdrawReportModel = exports.FeesDepositReportModel = exports.DepositReportModel = exports.EarningsReportModel = exports.DrvsAirdropReportModel = exports.PerpChangeLeverageReportModel = exports.LogType = void 0;
const auto_buffer_1 = require("./auto_buffer");
var LogType;
(function (LogType) {
    LogType[LogType["deposit"] = 1] = "deposit";
    LogType[LogType["withdraw"] = 2] = "withdraw";
    LogType[LogType["perpDeposit"] = 3] = "perpDeposit";
    LogType[LogType["perpWithdraw"] = 4] = "perpWithdraw";
    LogType[LogType["feesDeposit"] = 5] = "feesDeposit";
    LogType[LogType["feesWithdraw"] = 6] = "feesWithdraw";
    LogType[LogType["spotLpTrade"] = 7] = "spotLpTrade";
    LogType[LogType["earnings"] = 8] = "earnings";
    LogType[LogType["drvsAirdrop"] = 9] = "drvsAirdrop";
    LogType[LogType["spotPlaceOrder"] = 10] = "spotPlaceOrder";
    LogType[LogType["spotFillOrder"] = 11] = "spotFillOrder";
    LogType[LogType["spotNewOrder"] = 12] = "spotNewOrder";
    LogType[LogType["spotOrderCancel"] = 13] = "spotOrderCancel";
    LogType[LogType["spotOrderRevoke"] = 14] = "spotOrderRevoke";
    LogType[LogType["spotFees"] = 15] = "spotFees";
    LogType[LogType["spotPlaceMassCancel"] = 16] = "spotPlaceMassCancel";
    LogType[LogType["spotMassCancel"] = 17] = "spotMassCancel";
    LogType[LogType["perpPlaceOrder"] = 18] = "perpPlaceOrder";
    LogType[LogType["perpFillOrder"] = 19] = "perpFillOrder";
    LogType[LogType["perpNewOrder"] = 20] = "perpNewOrder";
    LogType[LogType["perpOrderCancel"] = 21] = "perpOrderCancel";
    LogType[LogType["perpOrderRevoke"] = 22] = "perpOrderRevoke";
    LogType[LogType["perpFees"] = 23] = "perpFees";
    LogType[LogType["perpFunding"] = 24] = "perpFunding";
    LogType[LogType["perpPlaceMassCancel"] = 25] = "perpPlaceMassCancel";
    LogType[LogType["perpMassCancel"] = 26] = "perpMassCancel";
    LogType[LogType["perpSocLoss"] = 27] = "perpSocLoss";
    LogType[LogType["perpChangeLeverage"] = 28] = "perpChangeLeverage";
    LogType[LogType["buyMarketSeat"] = 29] = "buyMarketSeat";
    LogType[LogType["sellMarketSeat"] = 30] = "sellMarketSeat";
    LogType[LogType["swapOrder"] = 31] = "swapOrder";
    LogType[LogType["moveSpot"] = 32] = "moveSpot";
    LogType[LogType["newPrivateClient"] = 33] = "newPrivateClient";
})(LogType || (exports.LogType = LogType = {}));
class PerpChangeLeverageReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpChangeLeverageReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.leverage = autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        return result;
    }
}
exports.PerpChangeLeverageReportModel = PerpChangeLeverageReportModel;
PerpChangeLeverageReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4; // 16 bytes
PerpChangeLeverageReportModel.OFFSET_TAG = 0;
PerpChangeLeverageReportModel.OFFSET_LEVERAGE = 1;
PerpChangeLeverageReportModel.OFFSET_CLIENT_ID = 4;
PerpChangeLeverageReportModel.OFFSET_INSTR_ID = 8;
PerpChangeLeverageReportModel.OFFSET_TIME = 12;
class DrvsAirdropReportModel {
    static fromBuffer(buffer, offset) {
        const result = new DrvsAirdropReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        result.time = autoBuffer.readU32();
        autoBuffer.readU32();
        return result;
    }
}
exports.DrvsAirdropReportModel = DrvsAirdropReportModel;
DrvsAirdropReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
DrvsAirdropReportModel.OFFSET_TAG = 0;
DrvsAirdropReportModel.OFFSET_CLIENT_ID = 4;
DrvsAirdropReportModel.OFFSET_AMOUNT = 8;
DrvsAirdropReportModel.OFFSET_TIME = 16;
class EarningsReportModel {
    static fromBuffer(buffer, offset) {
        const result = new EarningsReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.tokenId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        return result;
    }
}
exports.EarningsReportModel = EarningsReportModel;
EarningsReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
EarningsReportModel.OFFSET_TAG = 0;
EarningsReportModel.OFFSET_CLIENT_ID = 4;
EarningsReportModel.OFFSET_TOKEN_ID = 8;
EarningsReportModel.OFFSET_TIME = 12;
EarningsReportModel.OFFSET_AMOUNT = 16;
class DepositReportModel {
    static fromBuffer(buffer, offset) {
        const result = new DepositReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.tokenId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        return result;
    }
}
exports.DepositReportModel = DepositReportModel;
DepositReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
DepositReportModel.OFFSET_TAG = 0;
DepositReportModel.OFFSET_CLIENT_ID = 4;
DepositReportModel.OFFSET_TOKEN_ID = 8;
DepositReportModel.OFFSET_TIME = 12;
DepositReportModel.OFFSET_AMOUNT = 16;
class FeesDepositReportModel {
    static fromBuffer(buffer, offset) {
        const result = new FeesDepositReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.tokenId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        return result;
    }
}
exports.FeesDepositReportModel = FeesDepositReportModel;
FeesDepositReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
FeesDepositReportModel.OFFSET_TAG = 0;
FeesDepositReportModel.OFFSET_CLIENT_ID = 4;
FeesDepositReportModel.OFFSET_TOKEN_ID = 8;
FeesDepositReportModel.OFFSET_TIME = 12;
FeesDepositReportModel.OFFSET_AMOUNT = 16;
class FeesWithdrawReportModel {
    static fromBuffer(buffer, offset) {
        const result = new FeesWithdrawReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.tokenId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        return result;
    }
}
exports.FeesWithdrawReportModel = FeesWithdrawReportModel;
FeesWithdrawReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
FeesWithdrawReportModel.OFFSET_TAG = 0;
FeesWithdrawReportModel.OFFSET_CLIENT_ID = 4;
FeesWithdrawReportModel.OFFSET_TOKEN_ID = 8;
FeesWithdrawReportModel.OFFSET_TIME = 12;
FeesWithdrawReportModel.OFFSET_AMOUNT = 16;
class PerpDepositReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpDepositReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        return result;
    }
}
exports.PerpDepositReportModel = PerpDepositReportModel;
PerpDepositReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
PerpDepositReportModel.OFFSET_TAG = 0;
PerpDepositReportModel.OFFSET_CLIENT_ID = 4;
PerpDepositReportModel.OFFSET_INSTR_ID = 8;
PerpDepositReportModel.OFFSET_TIME = 12;
PerpDepositReportModel.OFFSET_AMOUNT = 16;
class BuyMarketSeatReportModel {
    static fromBuffer(buffer, offset) {
        const result = new BuyMarketSeatReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        result.seatPrice = autoBuffer.readI64();
        return result;
    }
}
exports.BuyMarketSeatReportModel = BuyMarketSeatReportModel;
BuyMarketSeatReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 2 * 8; // 32 bytes
BuyMarketSeatReportModel.OFFSET_TAG = 0;
BuyMarketSeatReportModel.OFFSET_CLIENT_ID = 4;
BuyMarketSeatReportModel.OFFSET_INSTR_ID = 8;
BuyMarketSeatReportModel.OFFSET_TIME = 12;
BuyMarketSeatReportModel.OFFSET_AMOUNT = 16;
BuyMarketSeatReportModel.OFFSET_SEAT_PRICE = 24;
class SellMarketSeatReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SellMarketSeatReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.seatPrice = autoBuffer.readI64();
        return result;
    }
}
exports.SellMarketSeatReportModel = SellMarketSeatReportModel;
SellMarketSeatReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
SellMarketSeatReportModel.OFFSET_TAG = 0;
SellMarketSeatReportModel.OFFSET_CLIENT_ID = 4;
SellMarketSeatReportModel.OFFSET_INSTR_ID = 8;
SellMarketSeatReportModel.OFFSET_TIME = 12;
SellMarketSeatReportModel.OFFSET_SEAT_PRICE = 16;
class WithdrawReportModel {
    static fromBuffer(buffer, offset) {
        const result = new WithdrawReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.tokenId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        return result;
    }
}
exports.WithdrawReportModel = WithdrawReportModel;
WithdrawReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
WithdrawReportModel.OFFSET_TAG = 0;
WithdrawReportModel.OFFSET_CLIENT_ID = 4;
WithdrawReportModel.OFFSET_TOKEN_ID = 8;
WithdrawReportModel.OFFSET_TIME = 12;
WithdrawReportModel.OFFSET_AMOUNT = 16;
class PerpWithdrawReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpWithdrawReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.amount = autoBuffer.readI64();
        return result;
    }
}
exports.PerpWithdrawReportModel = PerpWithdrawReportModel;
PerpWithdrawReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
PerpWithdrawReportModel.OFFSET_TAG = 0;
PerpWithdrawReportModel.OFFSET_CLIENT_ID = 4;
PerpWithdrawReportModel.OFFSET_INSTR_ID = 8;
PerpWithdrawReportModel.OFFSET_TIME = 12;
PerpWithdrawReportModel.OFFSET_AMOUNT = 16;
class SpotlpTradeReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotlpTradeReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.qty = autoBuffer.readI64();
        result.tokens = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.SpotlpTradeReportModel = SpotlpTradeReportModel;
SpotlpTradeReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 4 * 8; // 48 bytes
SpotlpTradeReportModel.OFFSET_TAG = 0;
SpotlpTradeReportModel.OFFSET_SIDE = 1;
SpotlpTradeReportModel.OFFSET_CLIENT_ID = 4;
SpotlpTradeReportModel.OFFSET_TIME = 8;
SpotlpTradeReportModel.OFFSET_INSTR_ID = 12;
SpotlpTradeReportModel.OFFSET_ORDER_ID = 16;
SpotlpTradeReportModel.OFFSET_QTY = 24;
SpotlpTradeReportModel.OFFSET_TOKENS = 32;
SpotlpTradeReportModel.OFFSET_CRNCY = 40;
class PerpFillOrderReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpFillOrderReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.perps = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        result.price = autoBuffer.readI64();
        result.rebates = autoBuffer.readI64();
        return result;
    }
}
exports.PerpFillOrderReportModel = PerpFillOrderReportModel;
PerpFillOrderReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 5 * 8; // 48 bytes
PerpFillOrderReportModel.OFFSET_TAG = 0;
PerpFillOrderReportModel.OFFSET_SIDE = 1;
PerpFillOrderReportModel.OFFSET_CLIENT_ID = 4;
PerpFillOrderReportModel.OFFSET_ORDER_ID = 8;
PerpFillOrderReportModel.OFFSET_PERPS = 16;
PerpFillOrderReportModel.OFFSET_CRNCY = 24;
PerpFillOrderReportModel.OFFSET_PRICE = 32;
PerpFillOrderReportModel.OFFSET_REBATES = 40;
class SpotFillOrderReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotFillOrderReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.qty = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        result.price = autoBuffer.readI64();
        result.rebates = autoBuffer.readI64();
        return result;
    }
}
exports.SpotFillOrderReportModel = SpotFillOrderReportModel;
SpotFillOrderReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 5 * 8; // 48 bytes
SpotFillOrderReportModel.OFFSET_TAG = 0;
SpotFillOrderReportModel.OFFSET_SIDE = 1;
SpotFillOrderReportModel.OFFSET_CLIENT_ID = 4;
SpotFillOrderReportModel.OFFSET_ORDER_ID = 8;
SpotFillOrderReportModel.OFFSET_QTY = 16;
SpotFillOrderReportModel.OFFSET_CRNCY = 24;
SpotFillOrderReportModel.OFFSET_PRICE = 32;
SpotFillOrderReportModel.OFFSET_REBATES = 40;
class PerpPlaceOrderReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpPlaceOrderReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.ioc = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        result.orderType = autoBuffer.readU8();
        result.clientId = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.perps = autoBuffer.readI64();
        result.price = autoBuffer.readI64();
        result.instrId = autoBuffer.readU32();
        result.leverage = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        autoBuffer.readU32();
        return result;
    }
}
exports.PerpPlaceOrderReportModel = PerpPlaceOrderReportModel;
PerpPlaceOrderReportModel.LENGTH = 4 * 1 + 5 * 4 + 3 * 8; // 48 bytes
PerpPlaceOrderReportModel.OFFSET_TAG = 0;
PerpPlaceOrderReportModel.OFFSET_IOC = 1;
PerpPlaceOrderReportModel.OFFSET_SIDE = 2;
PerpPlaceOrderReportModel.OFFSET_ORDER_TYPE = 3;
PerpPlaceOrderReportModel.OFFSET_CLIENT_ID = 4;
PerpPlaceOrderReportModel.OFFSET_ORDER_ID = 8;
PerpPlaceOrderReportModel.OFFSET_PERPS = 16;
PerpPlaceOrderReportModel.OFFSET_PRICE = 24;
PerpPlaceOrderReportModel.OFFSET_INSTR_ID = 32;
PerpPlaceOrderReportModel.OFFSET_LEVERAGE = 36;
PerpPlaceOrderReportModel.OFFSET_TIME = 40;
class SpotPlaceOrderReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotPlaceOrderReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.ioc = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        result.orderType = autoBuffer.readU8();
        result.clientId = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.qty = autoBuffer.readI64();
        result.price = autoBuffer.readI64();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        return result;
    }
}
exports.SpotPlaceOrderReportModel = SpotPlaceOrderReportModel;
SpotPlaceOrderReportModel.LENGTH = 4 * 1 + 3 * 4 + 3 * 8; // 40 bytes
SpotPlaceOrderReportModel.OFFSET_TAG = 0;
SpotPlaceOrderReportModel.OFFSET_IOC = 1;
SpotPlaceOrderReportModel.OFFSET_SIDE = 2;
SpotPlaceOrderReportModel.OFFSET_ORDER_TYPE = 3;
SpotPlaceOrderReportModel.OFFSET_CLIENT_ID = 4;
SpotPlaceOrderReportModel.OFFSET_ORDER_ID = 8;
SpotPlaceOrderReportModel.OFFSET_QTY = 16;
SpotPlaceOrderReportModel.OFFSET_PRICE = 24;
SpotPlaceOrderReportModel.OFFSET_INSTR_ID = 32;
SpotPlaceOrderReportModel.OFFSET_TIME = 36;
class SwapOrderReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SwapOrderReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        result.orderType = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.qty = autoBuffer.readI64();
        result.price = autoBuffer.readI64();
        result.time = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        return result;
    }
}
exports.SwapOrderReportModel = SwapOrderReportModel;
SwapOrderReportModel.LENGTH = 4 * 1 + 3 * 4 + 3 * 8; // 40 bytes
SwapOrderReportModel.OFFSET_TAG = 0;
SwapOrderReportModel.OFFSET_SIDE = 1;
SwapOrderReportModel.OFFSET_ORDER_TYPE = 2;
SwapOrderReportModel.OFFSET_ORDER_ID = 8;
SwapOrderReportModel.OFFSET_QTY = 16;
SwapOrderReportModel.OFFSET_PRICE = 24;
SwapOrderReportModel.OFFSET_TIME = 32;
SwapOrderReportModel.OFFSET_INSTR_ID = 36;
class PerpPlaceMassCancelReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpPlaceMassCancelReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        return result;
    }
}
exports.PerpPlaceMassCancelReportModel = PerpPlaceMassCancelReportModel;
PerpPlaceMassCancelReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4; // 16 bytes
PerpPlaceMassCancelReportModel.OFFSET_TAG = 0;
PerpPlaceMassCancelReportModel.OFFSET_CLIENT_ID = 4;
PerpPlaceMassCancelReportModel.OFFSET_INSTR_ID = 8;
PerpPlaceMassCancelReportModel.OFFSET_TIME = 12;
class SpotPlaceMassCancelReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotPlaceMassCancelReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        return result;
    }
}
exports.SpotPlaceMassCancelReportModel = SpotPlaceMassCancelReportModel;
SpotPlaceMassCancelReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4; // 16 bytes
SpotPlaceMassCancelReportModel.OFFSET_TAG = 0;
SpotPlaceMassCancelReportModel.OFFSET_CLIENT_ID = 4;
SpotPlaceMassCancelReportModel.OFFSET_INSTR_ID = 8;
SpotPlaceMassCancelReportModel.OFFSET_TIME = 12;
class PerpMassCancelReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpMassCancelReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.perps = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.PerpMassCancelReportModel = PerpMassCancelReportModel;
PerpMassCancelReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 3 * 8; // 32 bytes
PerpMassCancelReportModel.OFFSET_TAG = 0;
PerpMassCancelReportModel.OFFSET_SIDE = 1;
PerpMassCancelReportModel.OFFSET_ORDER_ID = 8;
PerpMassCancelReportModel.OFFSET_PERPS = 16;
PerpMassCancelReportModel.OFFSET_CRNCY = 24;
class SpotMassCancelReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotMassCancelReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.qty = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.SpotMassCancelReportModel = SpotMassCancelReportModel;
SpotMassCancelReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 3 * 8; // 32 bytes
SpotMassCancelReportModel.OFFSET_TAG = 0;
SpotMassCancelReportModel.OFFSET_SIDE = 1;
SpotMassCancelReportModel.OFFSET_ORDER_ID = 8;
SpotMassCancelReportModel.OFFSET_QTY = 16;
SpotMassCancelReportModel.OFFSET_CRNCY = 24;
class PerpFeesReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpFeesReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.refClientId = autoBuffer.readU32();
        result.fees = autoBuffer.readI64();
        result.refPayment = autoBuffer.readI64();
        return result;
    }
}
exports.PerpFeesReportModel = PerpFeesReportModel;
PerpFeesReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 2 * 8; // 24 bytes
PerpFeesReportModel.OFFSET_TAG = 0;
PerpFeesReportModel.OFFSET_REF_CLIENT_ID = 4;
PerpFeesReportModel.OFFSET_FEES = 8;
PerpFeesReportModel.OFFSET_REF_PAYMENT = 16;
class SpotFeesReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotFeesReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.refClientId = autoBuffer.readU32();
        result.fees = autoBuffer.readI64();
        result.refPayment = autoBuffer.readI64();
        return result;
    }
}
exports.SpotFeesReportModel = SpotFeesReportModel;
SpotFeesReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 2 * 8; // 24 bytes
SpotFeesReportModel.OFFSET_TAG = 0;
SpotFeesReportModel.OFFSET_REF_CLIENT_ID = 4;
SpotFeesReportModel.OFFSET_FEES = 8;
SpotFeesReportModel.OFFSET_REF_PAYMENT = 16;
class PerpFundingReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpFundingReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.funding = autoBuffer.readI64();
        return result;
    }
}
exports.PerpFundingReportModel = PerpFundingReportModel;
PerpFundingReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
PerpFundingReportModel.OFFSET_TAG = 0;
PerpFundingReportModel.OFFSET_CLIENT_ID = 4;
PerpFundingReportModel.OFFSET_INSTR_ID = 8;
PerpFundingReportModel.OFFSET_TIME = 12;
PerpFundingReportModel.OFFSET_FUNDING = 16;
class PerpSocLossReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpSocLossReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.socLoss = autoBuffer.readI64();
        return result;
    }
}
exports.PerpSocLossReportModel = PerpSocLossReportModel;
PerpSocLossReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 1 * 8; // 24 bytes
PerpSocLossReportModel.OFFSET_TAG = 0;
PerpSocLossReportModel.OFFSET_CLIENT_ID = 4;
PerpSocLossReportModel.OFFSET_INSTR_ID = 8;
PerpSocLossReportModel.OFFSET_TIME = 12;
PerpSocLossReportModel.OFFSET_SOC_LOSS = 16;
class PerpNewOrderReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpNewOrderReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        autoBuffer.readU32();
        result.perps = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.PerpNewOrderReportModel = PerpNewOrderReportModel;
PerpNewOrderReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 2 * 8; // 24 bytes
PerpNewOrderReportModel.OFFSET_TAG = 0;
PerpNewOrderReportModel.OFFSET_SIDE = 1;
PerpNewOrderReportModel.OFFSET_PERPS = 8;
PerpNewOrderReportModel.OFFSET_CRNCY = 16;
class SpotNewOrderReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotNewOrderReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        autoBuffer.readU32();
        result.qty = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.SpotNewOrderReportModel = SpotNewOrderReportModel;
SpotNewOrderReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 2 * 8; // 24 bytes
SpotNewOrderReportModel.OFFSET_TAG = 0;
SpotNewOrderReportModel.OFFSET_SIDE = 1;
SpotNewOrderReportModel.OFFSET_QTY = 8;
SpotNewOrderReportModel.OFFSET_CRNCY = 16;
class PerpOrderCancelReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpOrderCancelReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.perps = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.PerpOrderCancelReportModel = PerpOrderCancelReportModel;
PerpOrderCancelReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 3 * 8; // 40 bytes
PerpOrderCancelReportModel.OFFSET_TAG = 0;
PerpOrderCancelReportModel.OFFSET_SIDE = 1;
PerpOrderCancelReportModel.OFFSET_CLIENT_ID = 4;
PerpOrderCancelReportModel.OFFSET_INSTR_ID = 8;
PerpOrderCancelReportModel.OFFSET_TIME = 12;
PerpOrderCancelReportModel.OFFSET_ORDER_ID = 16;
PerpOrderCancelReportModel.OFFSET_PERPS = 24;
PerpOrderCancelReportModel.OFFSET_CRNCY = 32;
class SpotOrderCancelReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotOrderCancelReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.qty = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.SpotOrderCancelReportModel = SpotOrderCancelReportModel;
SpotOrderCancelReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 3 * 8; // 40 bytes
SpotOrderCancelReportModel.OFFSET_TAG = 0;
SpotOrderCancelReportModel.OFFSET_SIDE = 1;
SpotOrderCancelReportModel.OFFSET_CLIENT_ID = 4;
SpotOrderCancelReportModel.OFFSET_INSTR_ID = 8;
SpotOrderCancelReportModel.OFFSET_TIME = 12;
SpotOrderCancelReportModel.OFFSET_ORDER_ID = 16;
SpotOrderCancelReportModel.OFFSET_QTY = 24;
SpotOrderCancelReportModel.OFFSET_CRNCY = 32;
class PerpOrderRevokeReportModel {
    static fromBuffer(buffer, offset) {
        const result = new PerpOrderRevokeReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.perps = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.PerpOrderRevokeReportModel = PerpOrderRevokeReportModel;
PerpOrderRevokeReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 3 * 8; // 32 bytes
PerpOrderRevokeReportModel.OFFSET_TAG = 0;
PerpOrderRevokeReportModel.OFFSET_SIDE = 1;
PerpOrderRevokeReportModel.OFFSET_CLIENT_ID = 4;
PerpOrderRevokeReportModel.OFFSET_ORDER_ID = 8;
PerpOrderRevokeReportModel.OFFSET_PERPS = 16;
PerpOrderRevokeReportModel.OFFSET_CRNCY = 24;
class SpotOrderRevokeReportModel {
    static fromBuffer(buffer, offset) {
        const result = new SpotOrderRevokeReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        result.side = autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.orderId = autoBuffer.readI64();
        result.qty = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.SpotOrderRevokeReportModel = SpotOrderRevokeReportModel;
SpotOrderRevokeReportModel.LENGTH = 2 * 1 + 1 * 2 + 1 * 4 + 3 * 8; // 32 bytes
SpotOrderRevokeReportModel.OFFSET_TAG = 0;
SpotOrderRevokeReportModel.OFFSET_SIDE = 1;
SpotOrderRevokeReportModel.OFFSET_CLIENT_ID = 4;
SpotOrderRevokeReportModel.OFFSET_ORDER_ID = 8;
SpotOrderRevokeReportModel.OFFSET_QTY = 16;
SpotOrderRevokeReportModel.OFFSET_CRNCY = 24;
class MoveSpotAvailFundsReportModel {
    static fromBuffer(buffer, offset) {
        const result = new MoveSpotAvailFundsReportModel();
        let autoBuffer = new auto_buffer_1.AutoBuffer(buffer, offset);
        result.tag = autoBuffer.readU8();
        autoBuffer.readU8();
        autoBuffer.readU16();
        result.clientId = autoBuffer.readU32();
        result.instrId = autoBuffer.readU32();
        result.time = autoBuffer.readU32();
        result.qty = autoBuffer.readI64();
        result.crncy = autoBuffer.readI64();
        return result;
    }
}
exports.MoveSpotAvailFundsReportModel = MoveSpotAvailFundsReportModel;
MoveSpotAvailFundsReportModel.LENGTH = 2 * 1 + 1 * 2 + 3 * 4 + 2 * 8; // 32 bytes
MoveSpotAvailFundsReportModel.OFFSET_TAG = 0;
MoveSpotAvailFundsReportModel.OFFSET_CLIENT_ID = 4;
MoveSpotAvailFundsReportModel.OFFSET_INSTR_ID = 8;
MoveSpotAvailFundsReportModel.OFFSET_TIME = 12;
MoveSpotAvailFundsReportModel.OFFSET_QTY = 16;
MoveSpotAvailFundsReportModel.OFFSET_CRNCY = 24;
