// @deriverse/kit/dist/structure_models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivateClientModel = exports.PrivateClientHeaderModel = exports.PxOrdersModel = exports.OrderModel = exports.ClientPrimaryAccountHeaderModel = exports.PerpClientInfo5Model = exports.PerpClientInfo4Model = exports.PerpClientInfo3Model = exports.PerpClientInfo2Model = exports.PerpClientInfoModel = exports.SpotClientInfo2Model = exports.SpotClientInfoModel = exports.AssetRecordModel = exports.ClientPerpModel = exports.ClientSpotModel = exports.BaseCrncyRecordModel = exports.TokenStateModel = exports.RootStateModel = exports.InstrAccountHeaderModel = exports.CandlesAccountHeaderModel = exports.CandleModel = exports.LineQuotesModel = exports.OperatorModel = exports.HolderAccountHeaderModel = exports.PerpTradeAccountHeaderModel = exports.SpotTradeAccountHeaderModel = exports.CommunityAccountHeaderModel = exports.ClientCommunityAccountHeaderModel = exports.ClientCommunityRecordModel = exports.AccountType = exports.OrderType = void 0;
const auto_data_1 = require("./auto_data");
var OrderType;
(function (OrderType) {
    OrderType[OrderType["limit"] = 0] = "limit";
    OrderType[OrderType["market"] = 1] = "market";
    OrderType[OrderType["marginCall"] = 2] = "marginCall";
    OrderType[OrderType["forcedClose"] = 3] = "forcedClose";
})(OrderType || (exports.OrderType = OrderType = {}));
var AccountType;
(function (AccountType) {
    AccountType[AccountType["clientCommunity"] = 35] = "clientCommunity";
    AccountType[AccountType["clientPrimary"] = 31] = "clientPrimary";
    AccountType[AccountType["community"] = 34] = "community";
    AccountType[AccountType["futuresAskOrders"] = 29] = "futuresAskOrders";
    AccountType[AccountType["futuresAsksTree"] = 27] = "futuresAsksTree";
    AccountType[AccountType["futuresBidOrders"] = 28] = "futuresBidOrders";
    AccountType[AccountType["futuresBidsTree"] = 26] = "futuresBidsTree";
    AccountType[AccountType["futuresClientAccounts"] = 23] = "futuresClientAccounts";
    AccountType[AccountType["futuresClientInfos"] = 24] = "futuresClientInfos";
    AccountType[AccountType["futuresClientInfos2"] = 25] = "futuresClientInfos2";
    AccountType[AccountType["futuresLines"] = 30] = "futuresLines";
    AccountType[AccountType["futuresMaps"] = 22] = "futuresMaps";
    AccountType[AccountType["holder"] = 1] = "holder";
    AccountType[AccountType["root"] = 2] = "root";
    AccountType[AccountType["instr"] = 7] = "instr";
    AccountType[AccountType["spot15MCandles"] = 20] = "spot15MCandles";
    AccountType[AccountType["spot1MCandles"] = 19] = "spot1MCandles";
    AccountType[AccountType["spotAskOrders"] = 17] = "spotAskOrders";
    AccountType[AccountType["spotAsksTree"] = 15] = "spotAsksTree";
    AccountType[AccountType["spotBidOrders"] = 16] = "spotBidOrders";
    AccountType[AccountType["spotBidsTree"] = 14] = "spotBidsTree";
    AccountType[AccountType["spotClientInfos"] = 12] = "spotClientInfos";
    AccountType[AccountType["spotClientInfos2"] = 13] = "spotClientInfos2";
    AccountType[AccountType["spotDayCandles"] = 21] = "spotDayCandles";
    AccountType[AccountType["spotLines"] = 18] = "spotLines";
    AccountType[AccountType["spotMaps"] = 10] = "spotMaps";
    AccountType[AccountType["token"] = 4] = "token";
    AccountType[AccountType["perpAskOrders"] = 36] = "perpAskOrders";
    AccountType[AccountType["perpAsksTree"] = 37] = "perpAsksTree";
    AccountType[AccountType["perpBidOrders"] = 38] = "perpBidOrders";
    AccountType[AccountType["perpBidsTree"] = 39] = "perpBidsTree";
    AccountType[AccountType["perpClientInfos"] = 41] = "perpClientInfos";
    AccountType[AccountType["perpClientInfos2"] = 42] = "perpClientInfos2";
    AccountType[AccountType["perpClientInfos3"] = 43] = "perpClientInfos3";
    AccountType[AccountType["perpClientInfos4"] = 44] = "perpClientInfos4";
    AccountType[AccountType["perpClientInfos5"] = 45] = "perpClientInfos5";
    AccountType[AccountType["perpLines"] = 46] = "perpLines";
    AccountType[AccountType["perpMaps"] = 47] = "perpMaps";
    AccountType[AccountType["perpLongPxTree"] = 48] = "perpLongPxTree";
    AccountType[AccountType["perpShortPxTree"] = 49] = "perpShortPxTree";
    AccountType[AccountType["perpRebalanceTimeTree"] = 50] = "perpRebalanceTimeTree";
    AccountType[AccountType["privateClients"] = 51] = "privateClients";
})(AccountType || (exports.AccountType = AccountType = {}));
class ClientCommunityRecordModel {
    static fromBuffer(data, offset) {
        const result = new ClientCommunityRecordModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.dividendsRate = autoData.readF64();
        result.dividendsValue = autoData.readI64();
        result.feesPrepayment = autoData.readI64();
        result.feesRatio = autoData.readF64();
        result.refRewards = autoData.readI64();
        result.refPayments = autoData.readI64();
        result.lastFeesPrepaymentTime = autoData.readU32();
        result.crncyTokenId = autoData.readU32();
        return result;
    }
}
exports.ClientCommunityRecordModel = ClientCommunityRecordModel;
ClientCommunityRecordModel.LENGTH = 2 * 4 + 6 * 8; // 56 bytes
ClientCommunityRecordModel.OFFSET_DIVIDENDS_RATE = 0;
ClientCommunityRecordModel.OFFSET_DIVIDENDS_VALUE = 8;
ClientCommunityRecordModel.OFFSET_FEES_PREPAYMENT = 16;
ClientCommunityRecordModel.OFFSET_FEES_RATIO = 24;
ClientCommunityRecordModel.OFFSET_REF_REWARDS = 32;
ClientCommunityRecordModel.OFFSET_REF_PAYMENTS = 40;
ClientCommunityRecordModel.OFFSET_LAST_FEES_PREPAYMENT_TIME = 48;
ClientCommunityRecordModel.OFFSET_CRNCY_TOKEN_ID = 52;
class ClientCommunityAccountHeaderModel {
    static fromBuffer(data, offset) {
        const result = new ClientCommunityAccountHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.id = autoData.readU32();
        result.lastVotingTime = autoData.readU32();
        result.lastVotingCounter = autoData.readU32();
        result.currentVotingCounter = autoData.readU32();
        result.currentVotingTokens = autoData.readI64();
        result.lastVotingTokens = autoData.readI64();
        result.lastChoice = autoData.readU32();
        result.slot = autoData.readU32();
        result.drvsTokens = autoData.readI64();
        result.count = autoData.readU32();
        result.reserved = autoData.readU32();
        return result;
    }
}
exports.ClientCommunityAccountHeaderModel = ClientCommunityAccountHeaderModel;
ClientCommunityAccountHeaderModel.LENGTH = 10 * 4 + 3 * 8; // 64 bytes
ClientCommunityAccountHeaderModel.OFFSET_TAG = 0;
ClientCommunityAccountHeaderModel.OFFSET_VERSION = 4;
ClientCommunityAccountHeaderModel.OFFSET_ID = 8;
ClientCommunityAccountHeaderModel.OFFSET_LAST_VOTING_TIME = 12;
ClientCommunityAccountHeaderModel.OFFSET_LAST_VOTING_COUNTER = 16;
ClientCommunityAccountHeaderModel.OFFSET_CURRENT_VOTING_COUNTER = 20;
ClientCommunityAccountHeaderModel.OFFSET_CURRENT_VOTING_TOKENS = 24;
ClientCommunityAccountHeaderModel.OFFSET_LAST_VOTING_TOKENS = 32;
ClientCommunityAccountHeaderModel.OFFSET_LAST_CHOICE = 40;
ClientCommunityAccountHeaderModel.OFFSET_SLOT = 44;
ClientCommunityAccountHeaderModel.OFFSET_DRVS_TOKENS = 48;
ClientCommunityAccountHeaderModel.OFFSET_COUNT = 56;
ClientCommunityAccountHeaderModel.OFFSET_RESERVED = 60;
class CommunityAccountHeaderModel {
    static fromBuffer(data, offset) {
        const result = new CommunityAccountHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.drvsTokens = autoData.readI64();
        result.minAmount = autoData.readI64();
        result.votingSupply = autoData.readI64();
        result.prevVotingSupply = autoData.readI64();
        result.votingDecr = autoData.readI64();
        result.prevVotingDecr = autoData.readI64();
        result.votingUnchange = autoData.readI64();
        result.prevVotingUnchange = autoData.readI64();
        result.votingIncr = autoData.readI64();
        result.prevVotingIncr = autoData.readI64();
        result.votingCounter = autoData.readU32();
        result.votingStartSlot = autoData.readU32();
        result.votingEndTime = autoData.readU32();
        result.spotFeeRate = autoData.readU32();
        result.perpFeeRate = autoData.readU32();
        result.spotPoolRatio = autoData.readU32();
        result.marginCallPenaltyRate = autoData.readU32();
        result.feesPrepaymentForMaxDiscount = autoData.readU32();
        result.maxDiscount = autoData.readU32();
        result.reservedValue1 = autoData.readU32();
        result.reservedValue2 = autoData.readU32();
        result.reservedValue3 = autoData.readU32();
        result.reservedValue4 = autoData.readU32();
        result.reservedValue5 = autoData.readU32();
        result.reservedValue6 = autoData.readU32();
        result.reservedValue7 = autoData.readU32();
        result.reservedValue8 = autoData.readU32();
        result.count = autoData.readU32();
        return result;
    }
}
exports.CommunityAccountHeaderModel = CommunityAccountHeaderModel;
CommunityAccountHeaderModel.LENGTH = 20 * 4 + 10 * 8; // 160 bytes
CommunityAccountHeaderModel.OFFSET_TAG = 0;
CommunityAccountHeaderModel.OFFSET_VERSION = 4;
CommunityAccountHeaderModel.OFFSET_DRVS_TOKENS = 8;
CommunityAccountHeaderModel.OFFSET_MIN_AMOUNT = 16;
CommunityAccountHeaderModel.OFFSET_VOTING_SUPPLY = 24;
CommunityAccountHeaderModel.OFFSET_PREV_VOTING_SUPPLY = 32;
CommunityAccountHeaderModel.OFFSET_VOTING_DECR = 40;
CommunityAccountHeaderModel.OFFSET_PREV_VOTING_DECR = 48;
CommunityAccountHeaderModel.OFFSET_VOTING_UNCHANGE = 56;
CommunityAccountHeaderModel.OFFSET_PREV_VOTING_UNCHANGE = 64;
CommunityAccountHeaderModel.OFFSET_VOTING_INCR = 72;
CommunityAccountHeaderModel.OFFSET_PREV_VOTING_INCR = 80;
CommunityAccountHeaderModel.OFFSET_VOTING_COUNTER = 88;
CommunityAccountHeaderModel.OFFSET_VOTING_START_SLOT = 92;
CommunityAccountHeaderModel.OFFSET_VOTING_END_TIME = 96;
CommunityAccountHeaderModel.OFFSET_SPOT_FEE_RATE = 100;
CommunityAccountHeaderModel.OFFSET_PERP_FEE_RATE = 104;
CommunityAccountHeaderModel.OFFSET_SPOT_POOL_RATIO = 108;
CommunityAccountHeaderModel.OFFSET_MARGIN_CALL_PENALTY_RATE = 112;
CommunityAccountHeaderModel.OFFSET_FEES_PREPAYMENT_FOR_MAX_DISCOUNT = 116;
CommunityAccountHeaderModel.OFFSET_MAX_DISCOUNT = 120;
CommunityAccountHeaderModel.OFFSET_RESERVED_VALUE1 = 124;
CommunityAccountHeaderModel.OFFSET_RESERVED_VALUE2 = 128;
CommunityAccountHeaderModel.OFFSET_RESERVED_VALUE3 = 132;
CommunityAccountHeaderModel.OFFSET_RESERVED_VALUE4 = 136;
CommunityAccountHeaderModel.OFFSET_RESERVED_VALUE5 = 140;
CommunityAccountHeaderModel.OFFSET_RESERVED_VALUE6 = 144;
CommunityAccountHeaderModel.OFFSET_RESERVED_VALUE7 = 148;
CommunityAccountHeaderModel.OFFSET_RESERVED_VALUE8 = 152;
CommunityAccountHeaderModel.OFFSET_COUNT = 156;
class SpotTradeAccountHeaderModel {
    static fromBuffer(data, offset) {
        const result = new SpotTradeAccountHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.instrId = autoData.readU32();
        result.slot = autoData.readU32();
        result.assetTokenId = autoData.readU32();
        result.crncyTokenId = autoData.readU32();
        return result;
    }
}
exports.SpotTradeAccountHeaderModel = SpotTradeAccountHeaderModel;
SpotTradeAccountHeaderModel.LENGTH = 6 * 4; // 24 bytes
SpotTradeAccountHeaderModel.OFFSET_TAG = 0;
SpotTradeAccountHeaderModel.OFFSET_VERSION = 4;
SpotTradeAccountHeaderModel.OFFSET_INSTR_ID = 8;
SpotTradeAccountHeaderModel.OFFSET_SLOT = 12;
SpotTradeAccountHeaderModel.OFFSET_ASSET_TOKEN_ID = 16;
SpotTradeAccountHeaderModel.OFFSET_CRNCY_TOKEN_ID = 20;
class PerpTradeAccountHeaderModel {
    static fromBuffer(data, offset) {
        const result = new PerpTradeAccountHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.id = autoData.readU32();
        result.slot = autoData.readU32();
        result.assetTokenId = autoData.readU32();
        result.crncyTokenId = autoData.readU32();
        return result;
    }
}
exports.PerpTradeAccountHeaderModel = PerpTradeAccountHeaderModel;
PerpTradeAccountHeaderModel.LENGTH = 6 * 4; // 24 bytes
PerpTradeAccountHeaderModel.OFFSET_TAG = 0;
PerpTradeAccountHeaderModel.OFFSET_VERSION = 4;
PerpTradeAccountHeaderModel.OFFSET_ID = 8;
PerpTradeAccountHeaderModel.OFFSET_SLOT = 12;
PerpTradeAccountHeaderModel.OFFSET_ASSET_TOKEN_ID = 16;
PerpTradeAccountHeaderModel.OFFSET_CRNCY_TOKEN_ID = 20;
class HolderAccountHeaderModel {
    static fromBuffer(data, offset) {
        const result = new HolderAccountHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.operatorsCount = autoData.readU32();
        return result;
    }
}
exports.HolderAccountHeaderModel = HolderAccountHeaderModel;
HolderAccountHeaderModel.LENGTH = 2 * 4; // 8 bytes
HolderAccountHeaderModel.OFFSET_TAG = 0;
HolderAccountHeaderModel.OFFSET_OPERATORS_COUNT = 4;
class OperatorModel {
    static fromBuffer(data, offset) {
        const result = new OperatorModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.operatorAddress = autoData.readAddress();
        result.version = autoData.readU32();
        result.reserved = autoData.readU32();
        return result;
    }
}
exports.OperatorModel = OperatorModel;
OperatorModel.LENGTH = 2 * 4 + 1 * 32; // 40 bytes
OperatorModel.OFFSET_OPERATOR_ADDRESS = 0;
OperatorModel.OFFSET_VERSION = 32;
OperatorModel.OFFSET_RESERVED = 36;
class LineQuotesModel {
    static fromBuffer(data, offset) {
        const result = new LineQuotesModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.px = autoData.readI64();
        result.qty = autoData.readI64();
        return result;
    }
}
exports.LineQuotesModel = LineQuotesModel;
LineQuotesModel.LENGTH = 2 * 8; // 16 bytes
LineQuotesModel.OFFSET_PX = 0;
LineQuotesModel.OFFSET_QTY = 8;
class CandleModel {
    static fromBuffer(data, offset) {
        const result = new CandleModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.open = autoData.readI64();
        result.close = autoData.readI64();
        result.max = autoData.readI64();
        result.min = autoData.readI64();
        result.assetTokens = autoData.readI64();
        result.crncyTokens = autoData.readI64();
        result.time = autoData.readU32();
        result.counter = autoData.readU32();
        return result;
    }
}
exports.CandleModel = CandleModel;
CandleModel.LENGTH = 2 * 4 + 6 * 8; // 56 bytes
CandleModel.OFFSET_OPEN = 0;
CandleModel.OFFSET_CLOSE = 8;
CandleModel.OFFSET_MAX = 16;
CandleModel.OFFSET_MIN = 24;
CandleModel.OFFSET_ASSET_TOKENS = 32;
CandleModel.OFFSET_CRNCY_TOKENS = 40;
CandleModel.OFFSET_TIME = 48;
CandleModel.OFFSET_COUNTER = 52;
class CandlesAccountHeaderModel {
    static fromBuffer(data, offset) {
        const result = new CandlesAccountHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.id = autoData.readU32();
        result.slot = autoData.readU32();
        result.count = autoData.readU32();
        result.last = autoData.readU32();
        return result;
    }
}
exports.CandlesAccountHeaderModel = CandlesAccountHeaderModel;
CandlesAccountHeaderModel.LENGTH = 6 * 4; // 24 bytes
CandlesAccountHeaderModel.OFFSET_TAG = 0;
CandlesAccountHeaderModel.OFFSET_VERSION = 4;
CandlesAccountHeaderModel.OFFSET_ID = 8;
CandlesAccountHeaderModel.OFFSET_SLOT = 12;
CandlesAccountHeaderModel.OFFSET_COUNT = 16;
CandlesAccountHeaderModel.OFFSET_LAST = 20;
class InstrAccountHeaderModel {
    static fromBuffer(data, offset) {
        const result = new InstrAccountHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.instrId = autoData.readU32();
        result.assetTokenId = autoData.readU32();
        result.crncyTokenId = autoData.readU32();
        result.assetTokenDecsCount = autoData.readU32();
        result.crncyTokenDecsCount = autoData.readU32();
        result.mask = autoData.readU32();
        result.lastPx = autoData.readI64();
        result.lastClose = autoData.readI64();
        result.bestBid = autoData.readI64();
        result.bestAsk = autoData.readI64();
        result.perpLastPx = autoData.readI64();
        result.perpLastClose = autoData.readI64();
        result.perpOpenInt = autoData.readI64();
        result.variance = autoData.readF64();
        result.maxLeverage = autoData.readF64();
        result.prevDayTrades = autoData.readI64();
        result.perpInsuranceFund = autoData.readI64();
        result.perpPriceDelta = autoData.readF64();
        result.emaPx = autoData.readF64();
        result.lpPrevDayFees = autoData.readI64();
        result.assetTokens = autoData.readI64();
        result.crncyTokens = autoData.readI64();
        result.ps = autoData.readI64();
        result.poolFees = autoData.readI64();
        result.reservedValue1 = autoData.readI64();
        result.reservedValue2 = autoData.readI64();
        result.reservedValue3 = autoData.readI64();
        result.lastTradeAssetTokens = autoData.readI64();
        result.lastTradeCrncyTokens = autoData.readI64();
        result.dayLow = autoData.readI64();
        result.dayHigh = autoData.readI64();
        result.dayAssetTokens = autoData.readI64();
        result.dayCrncyTokens = autoData.readI64();
        result.perpClientsCount = autoData.readU32();
        result.perpDayTrades = autoData.readU32();
        result.perpBestBid = autoData.readI64();
        result.perpBestAsk = autoData.readI64();
        result.perpDayLow = autoData.readI64();
        result.perpDayHigh = autoData.readI64();
        result.perpDayAssetTokens = autoData.readI64();
        result.perpDayCrncyTokens = autoData.readI64();
        result.perpAlltimeTrades = autoData.readI64();
        result.perpPrevDayTrades = autoData.readI64();
        result.perpLastTradeAssetTokens = autoData.readI64();
        result.perpLastTradeCrncyTokens = autoData.readI64();
        result.alltimeTrades = autoData.readI64();
        result.dayVolatility = autoData.readF64();
        result.mapsAddress = autoData.readAddress();
        result.perpMapsAddress = autoData.readAddress();
        result.assetMint = autoData.readAddress();
        result.crncyMint = autoData.readAddress();
        result.lutAddress = autoData.readAddress();
        result.drvCount = autoData.readU32();
        result.slot = autoData.readU32();
        result.creator = autoData.readAddress();
        result.lastTime = autoData.readU32();
        result.distribTime = autoData.readU32();
        result.baseCrncyIndex = autoData.readU32();
        result.instanceCounter = autoData.readU32();
        result.varianceCounter = autoData.readU32();
        result.bidsTreeNodesCount = autoData.readU32();
        result.bidsTreeLinesEntry = autoData.readU32();
        result.bidsTreeOrdersEntry = autoData.readU32();
        result.asksTreeNodesCount = autoData.readU32();
        result.asksTreeLinesEntry = autoData.readU32();
        result.asksTreeOrdersEntry = autoData.readU32();
        result.bidLinesBegin = autoData.readU32();
        result.bidLinesEnd = autoData.readU32();
        result.bidLinesCount = autoData.readU32();
        result.askLinesBegin = autoData.readU32();
        result.askLinesEnd = autoData.readU32();
        result.askLinesCount = autoData.readU32();
        result.bidOrdersCount = autoData.readU32();
        result.askOrdersCount = autoData.readU32();
        result.fixingTime = autoData.readU32();
        result.fixingCrncyTokens = autoData.readI64();
        result.fixingAssetTokens = autoData.readI64();
        result.counter = autoData.readI64();
        result.protocolFees = autoData.readI64();
        result.hitsCounter = autoData.readI64();
        result.lastAssetTokens = autoData.readI64();
        result.lastCrncyTokens = autoData.readI64();
        result.perpUnderlyingPx = autoData.readI64();
        result.fixingPx = autoData.readI64();
        result.avgSpread = autoData.readF64();
        result.lastSpread = autoData.readF64();
        result.lastSpreadTime = autoData.readU32();
        result.totalSpreadPeriod = autoData.readU32();
        result.prevDayAssetTokens = autoData.readI64();
        result.prevDayCrncyTokens = autoData.readI64();
        result.alltimeAssetTokens = autoData.readF64();
        result.alltimeCrncyTokens = autoData.readF64();
        result.dayTrades = autoData.readU32();
        result.lpDayTrades = autoData.readU32();
        result.lpAlltimeFees = autoData.readF64();
        result.lpDayFees = autoData.readI64();
        result.lpPrevDayTrades = autoData.readU32();
        result.lpTime = autoData.readU32();
        result.feesTime = autoData.readU32();
        result.creationTime = autoData.readU32();
        result.decFactor = autoData.readI64();
        result.perpSlot = autoData.readU32();
        result.perpTime = autoData.readU32();
        result.perpFundingRateSlot = autoData.readU32();
        result.perpFundingRateTime = autoData.readU32();
        result.perpLongPxTreeNodesCount = autoData.readU32();
        result.perpLongPxTreeEntry = autoData.readU32();
        result.perpShortPxTreeNodesCount = autoData.readU32();
        result.perpShortPxTreeEntry = autoData.readU32();
        result.perpRebalanceTimeTreeNodesCount = autoData.readU32();
        result.perpRebalanceTimeTreeEntry = autoData.readU32();
        result.perpBidsTreeNodesCount = autoData.readU32();
        result.perpBidsTreeLinesEntry = autoData.readU32();
        result.perpBidsTreeOrdersEntry = autoData.readU32();
        result.perpAsksTreeNodesCount = autoData.readU32();
        result.perpAsksTreeLinesEntry = autoData.readU32();
        result.perpAsksTreeOrdersEntry = autoData.readU32();
        result.perpBidLinesBegin = autoData.readU32();
        result.perpBidLinesEnd = autoData.readU32();
        result.perpBidLinesCount = autoData.readU32();
        result.perpAskLinesBegin = autoData.readU32();
        result.perpAskLinesEnd = autoData.readU32();
        result.perpAskLinesCount = autoData.readU32();
        result.perpBidOrdersCount = autoData.readU32();
        result.perpAskOrdersCount = autoData.readU32();
        result.perpLongSpotPriceForWithdrowal = autoData.readI64();
        result.perpShortSpotPriceForWithdrowal = autoData.readI64();
        result.perpSocLossLongRate = autoData.readF64();
        result.perpSocLossShortRate = autoData.readF64();
        result.perpFundingRate = autoData.readF64();
        result.perpFundingFunds = autoData.readI64();
        result.perpSocLossFunds = autoData.readI64();
        result.perpPrevDayAssetTokens = autoData.readI64();
        result.perpPrevDayCrncyTokens = autoData.readI64();
        result.perpAlltimeAssetTokens = autoData.readF64();
        result.perpAlltimeCrncyTokens = autoData.readF64();
        result.liquidationThreshold = autoData.readF64();
        result.seatsReserve = autoData.readI64();
        result.reservedValue4 = autoData.readI64();
        result.reservedValue5 = autoData.readI64();
        result.reservedValue6 = autoData.readI64();
        result.reservedValue7 = autoData.readI64();
        result.reservedValue8 = autoData.readI64();
        result.reservedValue9 = autoData.readI64();
        result.reservedValue10 = autoData.readI64();
        return result;
    }
}
exports.InstrAccountHeaderModel = InstrAccountHeaderModel;
InstrAccountHeaderModel.LENGTH = 64 * 4 + 77 * 8 + 6 * 32; // 1064 bytes
InstrAccountHeaderModel.OFFSET_TAG = 0;
InstrAccountHeaderModel.OFFSET_VERSION = 4;
InstrAccountHeaderModel.OFFSET_INSTR_ID = 8;
InstrAccountHeaderModel.OFFSET_ASSET_TOKEN_ID = 12;
InstrAccountHeaderModel.OFFSET_CRNCY_TOKEN_ID = 16;
InstrAccountHeaderModel.OFFSET_ASSET_TOKEN_DECS_COUNT = 20;
InstrAccountHeaderModel.OFFSET_CRNCY_TOKEN_DECS_COUNT = 24;
InstrAccountHeaderModel.OFFSET_MASK = 28;
InstrAccountHeaderModel.OFFSET_LAST_PX = 32;
InstrAccountHeaderModel.OFFSET_LAST_CLOSE = 40;
InstrAccountHeaderModel.OFFSET_BEST_BID = 48;
InstrAccountHeaderModel.OFFSET_BEST_ASK = 56;
InstrAccountHeaderModel.OFFSET_PERP_LAST_PX = 64;
InstrAccountHeaderModel.OFFSET_PERP_LAST_CLOSE = 72;
InstrAccountHeaderModel.OFFSET_PERP_OPEN_INT = 80;
InstrAccountHeaderModel.OFFSET_VARIANCE = 88;
InstrAccountHeaderModel.OFFSET_MAX_LEVERAGE = 96;
InstrAccountHeaderModel.OFFSET_PREV_DAY_TRADES = 104;
InstrAccountHeaderModel.OFFSET_PERP_INSURANCE_FUND = 112;
InstrAccountHeaderModel.OFFSET_PERP_PRICE_DELTA = 120;
InstrAccountHeaderModel.OFFSET_EMA_PX = 128;
InstrAccountHeaderModel.OFFSET_LP_PREV_DAY_FEES = 136;
InstrAccountHeaderModel.OFFSET_ASSET_TOKENS = 144;
InstrAccountHeaderModel.OFFSET_CRNCY_TOKENS = 152;
InstrAccountHeaderModel.OFFSET_PS = 160;
InstrAccountHeaderModel.OFFSET_POOL_FEES = 168;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE1 = 176;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE2 = 184;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE3 = 192;
InstrAccountHeaderModel.OFFSET_LAST_TRADE_ASSET_TOKENS = 200;
InstrAccountHeaderModel.OFFSET_LAST_TRADE_CRNCY_TOKENS = 208;
InstrAccountHeaderModel.OFFSET_DAY_LOW = 216;
InstrAccountHeaderModel.OFFSET_DAY_HIGH = 224;
InstrAccountHeaderModel.OFFSET_DAY_ASSET_TOKENS = 232;
InstrAccountHeaderModel.OFFSET_DAY_CRNCY_TOKENS = 240;
InstrAccountHeaderModel.OFFSET_PERP_CLIENTS_COUNT = 248;
InstrAccountHeaderModel.OFFSET_PERP_DAY_TRADES = 252;
InstrAccountHeaderModel.OFFSET_PERP_BEST_BID = 256;
InstrAccountHeaderModel.OFFSET_PERP_BEST_ASK = 264;
InstrAccountHeaderModel.OFFSET_PERP_DAY_LOW = 272;
InstrAccountHeaderModel.OFFSET_PERP_DAY_HIGH = 280;
InstrAccountHeaderModel.OFFSET_PERP_DAY_ASSET_TOKENS = 288;
InstrAccountHeaderModel.OFFSET_PERP_DAY_CRNCY_TOKENS = 296;
InstrAccountHeaderModel.OFFSET_PERP_ALLTIME_TRADES = 304;
InstrAccountHeaderModel.OFFSET_PERP_PREV_DAY_TRADES = 312;
InstrAccountHeaderModel.OFFSET_PERP_LAST_TRADE_ASSET_TOKENS = 320;
InstrAccountHeaderModel.OFFSET_PERP_LAST_TRADE_CRNCY_TOKENS = 328;
InstrAccountHeaderModel.OFFSET_ALLTIME_TRADES = 336;
InstrAccountHeaderModel.OFFSET_DAY_VOLATILITY = 344;
InstrAccountHeaderModel.OFFSET_MAPS_ADDRESS = 352;
InstrAccountHeaderModel.OFFSET_PERP_MAPS_ADDRESS = 384;
InstrAccountHeaderModel.OFFSET_ASSET_MINT = 416;
InstrAccountHeaderModel.OFFSET_CRNCY_MINT = 448;
InstrAccountHeaderModel.OFFSET_LUT_ADDRESS = 480;
InstrAccountHeaderModel.OFFSET_DRV_COUNT = 512;
InstrAccountHeaderModel.OFFSET_SLOT = 516;
InstrAccountHeaderModel.OFFSET_CREATOR = 520;
InstrAccountHeaderModel.OFFSET_LAST_TIME = 552;
InstrAccountHeaderModel.OFFSET_DISTRIB_TIME = 556;
InstrAccountHeaderModel.OFFSET_BASE_CRNCY_INDEX = 560;
InstrAccountHeaderModel.OFFSET_INSTANCE_COUNTER = 564;
InstrAccountHeaderModel.OFFSET_VARIANCE_COUNTER = 568;
InstrAccountHeaderModel.OFFSET_BIDS_TREE_NODES_COUNT = 572;
InstrAccountHeaderModel.OFFSET_BIDS_TREE_LINES_ENTRY = 576;
InstrAccountHeaderModel.OFFSET_BIDS_TREE_ORDERS_ENTRY = 580;
InstrAccountHeaderModel.OFFSET_ASKS_TREE_NODES_COUNT = 584;
InstrAccountHeaderModel.OFFSET_ASKS_TREE_LINES_ENTRY = 588;
InstrAccountHeaderModel.OFFSET_ASKS_TREE_ORDERS_ENTRY = 592;
InstrAccountHeaderModel.OFFSET_BID_LINES_BEGIN = 596;
InstrAccountHeaderModel.OFFSET_BID_LINES_END = 600;
InstrAccountHeaderModel.OFFSET_BID_LINES_COUNT = 604;
InstrAccountHeaderModel.OFFSET_ASK_LINES_BEGIN = 608;
InstrAccountHeaderModel.OFFSET_ASK_LINES_END = 612;
InstrAccountHeaderModel.OFFSET_ASK_LINES_COUNT = 616;
InstrAccountHeaderModel.OFFSET_BID_ORDERS_COUNT = 620;
InstrAccountHeaderModel.OFFSET_ASK_ORDERS_COUNT = 624;
InstrAccountHeaderModel.OFFSET_FIXING_TIME = 628;
InstrAccountHeaderModel.OFFSET_FIXING_CRNCY_TOKENS = 632;
InstrAccountHeaderModel.OFFSET_FIXING_ASSET_TOKENS = 640;
InstrAccountHeaderModel.OFFSET_COUNTER = 648;
InstrAccountHeaderModel.OFFSET_PROTOCOL_FEES = 656;
InstrAccountHeaderModel.OFFSET_HITS_COUNTER = 664;
InstrAccountHeaderModel.OFFSET_LAST_ASSET_TOKENS = 672;
InstrAccountHeaderModel.OFFSET_LAST_CRNCY_TOKENS = 680;
InstrAccountHeaderModel.OFFSET_PERP_UNDERLYING_PX = 688;
InstrAccountHeaderModel.OFFSET_FIXING_PX = 696;
InstrAccountHeaderModel.OFFSET_AVG_SPREAD = 704;
InstrAccountHeaderModel.OFFSET_LAST_SPREAD = 712;
InstrAccountHeaderModel.OFFSET_LAST_SPREAD_TIME = 720;
InstrAccountHeaderModel.OFFSET_TOTAL_SPREAD_PERIOD = 724;
InstrAccountHeaderModel.OFFSET_PREV_DAY_ASSET_TOKENS = 728;
InstrAccountHeaderModel.OFFSET_PREV_DAY_CRNCY_TOKENS = 736;
InstrAccountHeaderModel.OFFSET_ALLTIME_ASSET_TOKENS = 744;
InstrAccountHeaderModel.OFFSET_ALLTIME_CRNCY_TOKENS = 752;
InstrAccountHeaderModel.OFFSET_DAY_TRADES = 760;
InstrAccountHeaderModel.OFFSET_LP_DAY_TRADES = 764;
InstrAccountHeaderModel.OFFSET_LP_ALLTIME_FEES = 768;
InstrAccountHeaderModel.OFFSET_LP_DAY_FEES = 776;
InstrAccountHeaderModel.OFFSET_LP_PREV_DAY_TRADES = 784;
InstrAccountHeaderModel.OFFSET_LP_TIME = 788;
InstrAccountHeaderModel.OFFSET_FEES_TIME = 792;
InstrAccountHeaderModel.OFFSET_CREATION_TIME = 796;
InstrAccountHeaderModel.OFFSET_DEC_FACTOR = 800;
InstrAccountHeaderModel.OFFSET_PERP_SLOT = 808;
InstrAccountHeaderModel.OFFSET_PERP_TIME = 812;
InstrAccountHeaderModel.OFFSET_PERP_FUNDING_RATE_SLOT = 816;
InstrAccountHeaderModel.OFFSET_PERP_FUNDING_RATE_TIME = 820;
InstrAccountHeaderModel.OFFSET_PERP_LONG_PX_TREE_NODES_COUNT = 824;
InstrAccountHeaderModel.OFFSET_PERP_LONG_PX_TREE_ENTRY = 828;
InstrAccountHeaderModel.OFFSET_PERP_SHORT_PX_TREE_NODES_COUNT = 832;
InstrAccountHeaderModel.OFFSET_PERP_SHORT_PX_TREE_ENTRY = 836;
InstrAccountHeaderModel.OFFSET_PERP_REBALANCE_TIME_TREE_NODES_COUNT = 840;
InstrAccountHeaderModel.OFFSET_PERP_REBALANCE_TIME_TREE_ENTRY = 844;
InstrAccountHeaderModel.OFFSET_PERP_BIDS_TREE_NODES_COUNT = 848;
InstrAccountHeaderModel.OFFSET_PERP_BIDS_TREE_LINES_ENTRY = 852;
InstrAccountHeaderModel.OFFSET_PERP_BIDS_TREE_ORDERS_ENTRY = 856;
InstrAccountHeaderModel.OFFSET_PERP_ASKS_TREE_NODES_COUNT = 860;
InstrAccountHeaderModel.OFFSET_PERP_ASKS_TREE_LINES_ENTRY = 864;
InstrAccountHeaderModel.OFFSET_PERP_ASKS_TREE_ORDERS_ENTRY = 868;
InstrAccountHeaderModel.OFFSET_PERP_BID_LINES_BEGIN = 872;
InstrAccountHeaderModel.OFFSET_PERP_BID_LINES_END = 876;
InstrAccountHeaderModel.OFFSET_PERP_BID_LINES_COUNT = 880;
InstrAccountHeaderModel.OFFSET_PERP_ASK_LINES_BEGIN = 884;
InstrAccountHeaderModel.OFFSET_PERP_ASK_LINES_END = 888;
InstrAccountHeaderModel.OFFSET_PERP_ASK_LINES_COUNT = 892;
InstrAccountHeaderModel.OFFSET_PERP_BID_ORDERS_COUNT = 896;
InstrAccountHeaderModel.OFFSET_PERP_ASK_ORDERS_COUNT = 900;
InstrAccountHeaderModel.OFFSET_PERP_LONG_SPOT_PRICE_FOR_WITHDROWAL = 904;
InstrAccountHeaderModel.OFFSET_PERP_SHORT_SPOT_PRICE_FOR_WITHDROWAL = 912;
InstrAccountHeaderModel.OFFSET_PERP_SOC_LOSS_LONG_RATE = 920;
InstrAccountHeaderModel.OFFSET_PERP_SOC_LOSS_SHORT_RATE = 928;
InstrAccountHeaderModel.OFFSET_PERP_FUNDING_RATE = 936;
InstrAccountHeaderModel.OFFSET_PERP_FUNDING_FUNDS = 944;
InstrAccountHeaderModel.OFFSET_PERP_SOC_LOSS_FUNDS = 952;
InstrAccountHeaderModel.OFFSET_PERP_PREV_DAY_ASSET_TOKENS = 960;
InstrAccountHeaderModel.OFFSET_PERP_PREV_DAY_CRNCY_TOKENS = 968;
InstrAccountHeaderModel.OFFSET_PERP_ALLTIME_ASSET_TOKENS = 976;
InstrAccountHeaderModel.OFFSET_PERP_ALLTIME_CRNCY_TOKENS = 984;
InstrAccountHeaderModel.OFFSET_LIQUIDATION_THRESHOLD = 992;
InstrAccountHeaderModel.OFFSET_SEATS_RESERVE = 1000;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE4 = 1008;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE5 = 1016;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE6 = 1024;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE7 = 1032;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE8 = 1040;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE9 = 1048;
InstrAccountHeaderModel.OFFSET_RESERVED_VALUE10 = 1056;
class RootStateModel {
    static fromBuffer(data, offset) {
        const result = new RootStateModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.operatorAddress = autoData.readAddress();
        result.holderAddress = autoData.readAddress();
        result.drvsMintAddress = autoData.readAddress();
        result.lutAddress = autoData.readAddress();
        result.airdropAuthorityAddress = autoData.readAddress();
        result.privateModeAuthorityAddress = autoData.readAddress();
        result.refProgramDuration = autoData.readU32();
        result.refLinkDuration = autoData.readU32();
        result.refDiscount = autoData.readF64();
        result.refRatio = autoData.readF64();
        result.clientsCount = autoData.readU32();
        result.tokensCount = autoData.readU32();
        result.instrCount = autoData.readU32();
        result.refCounter = autoData.readU32();
        result.mask = autoData.readU32();
        result.pointsProgramExpiration = autoData.readU32();
        result.purchasingPerpSeatFee = autoData.readF64();
        return result;
    }
}
exports.RootStateModel = RootStateModel;
RootStateModel.LENGTH = 10 * 4 + 3 * 8 + 6 * 32; // 256 bytes
RootStateModel.OFFSET_TAG = 0;
RootStateModel.OFFSET_VERSION = 4;
RootStateModel.OFFSET_OPERATOR_ADDRESS = 8;
RootStateModel.OFFSET_HOLDER_ADDRESS = 40;
RootStateModel.OFFSET_DRVS_MINT_ADDRESS = 72;
RootStateModel.OFFSET_LUT_ADDRESS = 104;
RootStateModel.OFFSET_AIRDROP_AUTHORITY_ADDRESS = 136;
RootStateModel.OFFSET_PRIVATE_MODE_AUTHORITY_ADDRESS = 168;
RootStateModel.OFFSET_REF_PROGRAM_DURATION = 200;
RootStateModel.OFFSET_REF_LINK_DURATION = 204;
RootStateModel.OFFSET_REF_DISCOUNT = 208;
RootStateModel.OFFSET_REF_RATIO = 216;
RootStateModel.OFFSET_CLIENTS_COUNT = 224;
RootStateModel.OFFSET_TOKENS_COUNT = 228;
RootStateModel.OFFSET_INSTR_COUNT = 232;
RootStateModel.OFFSET_REF_COUNTER = 236;
RootStateModel.OFFSET_MASK = 240;
RootStateModel.OFFSET_POINTS_PROGRAM_EXPIRATION = 244;
RootStateModel.OFFSET_PURCHASING_PERP_SEAT_FEE = 248;
class TokenStateModel {
    static fromBuffer(data, offset) {
        const result = new TokenStateModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.address = autoData.readAddress();
        result.programAddress = autoData.readAddress();
        result.id = autoData.readU32();
        result.mask = autoData.readU32();
        result.reserved = autoData.readU32();
        result.baseCrncyIndex = autoData.readU32();
        return result;
    }
}
exports.TokenStateModel = TokenStateModel;
TokenStateModel.LENGTH = 6 * 4 + 2 * 32; // 88 bytes
TokenStateModel.OFFSET_TAG = 0;
TokenStateModel.OFFSET_VERSION = 4;
TokenStateModel.OFFSET_ADDRESS = 8;
TokenStateModel.OFFSET_PROGRAM_ADDRESS = 40;
TokenStateModel.OFFSET_ID = 72;
TokenStateModel.OFFSET_MASK = 76;
TokenStateModel.OFFSET_RESERVED = 80;
TokenStateModel.OFFSET_BASE_CRNCY_INDEX = 84;
class BaseCrncyRecordModel {
    static fromBuffer(data, offset) {
        const result = new BaseCrncyRecordModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.crncyTokenId = autoData.readU32();
        result.decsCount = autoData.readU32();
        result.funds = autoData.readI64();
        result.rate = autoData.readF64();
        result.denominator = autoData.readF64();
        result.lockedDrvsAmount = autoData.readI64();
        result.lockedDrvsDividendsValue = autoData.readI64();
        result.mask = autoData.readI64();
        return result;
    }
}
exports.BaseCrncyRecordModel = BaseCrncyRecordModel;
BaseCrncyRecordModel.LENGTH = 2 * 4 + 6 * 8; // 56 bytes
BaseCrncyRecordModel.OFFSET_CRNCY_TOKEN_ID = 0;
BaseCrncyRecordModel.OFFSET_DECS_COUNT = 4;
BaseCrncyRecordModel.OFFSET_FUNDS = 8;
BaseCrncyRecordModel.OFFSET_RATE = 16;
BaseCrncyRecordModel.OFFSET_DENOMINATOR = 24;
BaseCrncyRecordModel.OFFSET_LOCKED_DRVS_AMOUNT = 32;
BaseCrncyRecordModel.OFFSET_LOCKED_DRVS_DIVIDENDS_VALUE = 40;
BaseCrncyRecordModel.OFFSET_MASK = 48;
class ClientSpotModel {
    static fromBuffer(data, offset) {
        const result = new ClientSpotModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.assetId = autoData.readU32();
        result.tempClientId = autoData.readU32();
        result.slot = autoData.readU32();
        result.padding = autoData.readU32();
        return result;
    }
}
exports.ClientSpotModel = ClientSpotModel;
ClientSpotModel.LENGTH = 4 * 4; // 16 bytes
ClientSpotModel.OFFSET_ASSET_ID = 0;
ClientSpotModel.OFFSET_TEMP_CLIENT_ID = 4;
ClientSpotModel.OFFSET_SLOT = 8;
ClientSpotModel.OFFSET_PADDING = 12;
class ClientPerpModel {
    static fromBuffer(data, offset) {
        const result = new ClientPerpModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.assetId = autoData.readU32();
        result.tempClientId = autoData.readU32();
        result.slot = autoData.readU32();
        result.padding = autoData.readU32();
        return result;
    }
}
exports.ClientPerpModel = ClientPerpModel;
ClientPerpModel.LENGTH = 4 * 4; // 16 bytes
ClientPerpModel.OFFSET_ASSET_ID = 0;
ClientPerpModel.OFFSET_TEMP_CLIENT_ID = 4;
ClientPerpModel.OFFSET_SLOT = 8;
ClientPerpModel.OFFSET_PADDING = 12;
class AssetRecordModel {
    static fromBuffer(data, offset) {
        const result = new AssetRecordModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.assetId = autoData.readU32();
        result.tempId = autoData.readU32();
        result.value = autoData.readI64();
        return result;
    }
}
exports.AssetRecordModel = AssetRecordModel;
AssetRecordModel.LENGTH = 2 * 4 + 1 * 8; // 16 bytes
AssetRecordModel.OFFSET_ASSET_ID = 0;
AssetRecordModel.OFFSET_TEMP_ID = 4;
AssetRecordModel.OFFSET_VALUE = 8;
class SpotClientInfoModel {
    static fromBuffer(data, offset) {
        const result = new SpotClientInfoModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.client = autoData.readU32();
        result.filledOrders = autoData.readU32();
        result.bidsEntry = autoData.readU32();
        result.asksEntry = autoData.readU32();
        result.availAssetTokens = autoData.readI64();
        result.availCrncyTokens = autoData.readI64();
        return result;
    }
}
exports.SpotClientInfoModel = SpotClientInfoModel;
SpotClientInfoModel.LENGTH = 4 * 4 + 2 * 8; // 32 bytes
SpotClientInfoModel.OFFSET_CLIENT = 0;
SpotClientInfoModel.OFFSET_FILLED_ORDERS = 4;
SpotClientInfoModel.OFFSET_BIDS_ENTRY = 8;
SpotClientInfoModel.OFFSET_ASKS_ENTRY = 12;
SpotClientInfoModel.OFFSET_AVAIL_ASSET_TOKENS = 16;
SpotClientInfoModel.OFFSET_AVAIL_CRNCY_TOKENS = 24;
class SpotClientInfo2Model {
    static fromBuffer(data, offset) {
        const result = new SpotClientInfo2Model();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.inOrdersAssetTokens = autoData.readI64();
        result.inOrdersCrncyTokens = autoData.readI64();
        result.bidSlot = autoData.readU32();
        result.askSlot = autoData.readU32();
        result.reserved = autoData.readI64();
        return result;
    }
}
exports.SpotClientInfo2Model = SpotClientInfo2Model;
SpotClientInfo2Model.LENGTH = 2 * 4 + 3 * 8; // 32 bytes
SpotClientInfo2Model.OFFSET_IN_ORDERS_ASSET_TOKENS = 0;
SpotClientInfo2Model.OFFSET_IN_ORDERS_CRNCY_TOKENS = 8;
SpotClientInfo2Model.OFFSET_BID_SLOT = 16;
SpotClientInfo2Model.OFFSET_ASK_SLOT = 20;
SpotClientInfo2Model.OFFSET_RESERVED = 24;
class PerpClientInfoModel {
    static fromBuffer(data, offset) {
        const result = new PerpClientInfoModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.funds = autoData.readI64();
        result.perps = autoData.readI64();
        result.inOrdersFunds = autoData.readI64();
        result.inOrdersPerps = autoData.readI64();
        return result;
    }
}
exports.PerpClientInfoModel = PerpClientInfoModel;
PerpClientInfoModel.LENGTH = 4 * 8; // 32 bytes
PerpClientInfoModel.OFFSET_FUNDS = 0;
PerpClientInfoModel.OFFSET_PERPS = 8;
PerpClientInfoModel.OFFSET_IN_ORDERS_FUNDS = 16;
PerpClientInfoModel.OFFSET_IN_ORDERS_PERPS = 24;
class PerpClientInfo2Model {
    static fromBuffer(data, offset) {
        const result = new PerpClientInfo2Model();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.cost = autoData.readI64();
        result.result = autoData.readI64();
        result.bidSlot = autoData.readU32();
        result.askSlot = autoData.readU32();
        result.pxNode = autoData.readU32();
        result.mask = autoData.readU32();
        return result;
    }
}
exports.PerpClientInfo2Model = PerpClientInfo2Model;
PerpClientInfo2Model.LENGTH = 4 * 4 + 2 * 8; // 32 bytes
PerpClientInfo2Model.OFFSET_COST = 0;
PerpClientInfo2Model.OFFSET_RESULT = 8;
PerpClientInfo2Model.OFFSET_BID_SLOT = 16;
PerpClientInfo2Model.OFFSET_ASK_SLOT = 20;
PerpClientInfo2Model.OFFSET_PX_NODE = 24;
PerpClientInfo2Model.OFFSET_MASK = 28;
class PerpClientInfo3Model {
    static fromBuffer(data, offset) {
        const result = new PerpClientInfo3Model();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.client = autoData.readU32();
        result.filledOrders = autoData.readU32();
        result.bidsEntry = autoData.readU32();
        result.asksEntry = autoData.readU32();
        result.fees = autoData.readI64();
        result.rebates = autoData.readI64();
        return result;
    }
}
exports.PerpClientInfo3Model = PerpClientInfo3Model;
PerpClientInfo3Model.LENGTH = 4 * 4 + 2 * 8; // 32 bytes
PerpClientInfo3Model.OFFSET_CLIENT = 0;
PerpClientInfo3Model.OFFSET_FILLED_ORDERS = 4;
PerpClientInfo3Model.OFFSET_BIDS_ENTRY = 8;
PerpClientInfo3Model.OFFSET_ASKS_ENTRY = 12;
PerpClientInfo3Model.OFFSET_FEES = 16;
PerpClientInfo3Model.OFFSET_REBATES = 24;
class PerpClientInfo4Model {
    static fromBuffer(data, offset) {
        const result = new PerpClientInfo4Model();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.lastSocLossRate = autoData.readF64();
        result.lastSocLossPerps = autoData.readI64();
        result.socLossFunds = autoData.readI64();
        result.lossCoverage = autoData.readI64();
        return result;
    }
}
exports.PerpClientInfo4Model = PerpClientInfo4Model;
PerpClientInfo4Model.LENGTH = 4 * 8; // 32 bytes
PerpClientInfo4Model.OFFSET_LAST_SOC_LOSS_RATE = 0;
PerpClientInfo4Model.OFFSET_LAST_SOC_LOSS_PERPS = 8;
PerpClientInfo4Model.OFFSET_SOC_LOSS_FUNDS = 16;
PerpClientInfo4Model.OFFSET_LOSS_COVERAGE = 24;
class PerpClientInfo5Model {
    static fromBuffer(data, offset) {
        const result = new PerpClientInfo5Model();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.fundingFunds = autoData.readI64();
        result.lastFundingRate = autoData.readF64();
        result.reserved = autoData.readI64();
        result.rebalanceTime = autoData.readU32();
        result.fundingNode = autoData.readU32();
        return result;
    }
}
exports.PerpClientInfo5Model = PerpClientInfo5Model;
PerpClientInfo5Model.LENGTH = 2 * 4 + 3 * 8; // 32 bytes
PerpClientInfo5Model.OFFSET_FUNDING_FUNDS = 0;
PerpClientInfo5Model.OFFSET_LAST_FUNDING_RATE = 8;
PerpClientInfo5Model.OFFSET_RESERVED = 16;
PerpClientInfo5Model.OFFSET_REBALANCE_TIME = 24;
PerpClientInfo5Model.OFFSET_FUNDING_NODE = 28;
class ClientPrimaryAccountHeaderModel {
    static fromBuffer(data, offset) {
        const result = new ClientPrimaryAccountHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        result.walletAddress = autoData.readAddress();
        result.lutAddress = autoData.readAddress();
        result.refAddress = autoData.readAddress();
        result.vmWalletAddress = autoData.readAddress();
        result.vmInstr0 = autoData.readU32();
        result.vmInstr1 = autoData.readU32();
        result.vmInstr2 = autoData.readU32();
        result.vmInstr3 = autoData.readU32();
        result.vmInstr4 = autoData.readU32();
        result.vmInstr5 = autoData.readU32();
        result.vmInstr6 = autoData.readU32();
        result.vmInstr7 = autoData.readU32();
        result.vmWithdrawTokenId = autoData.readU32();
        result.vmMask = autoData.readU32();
        result.vmWithdrawAmount = autoData.readI64();
        result.firstRefLinkDiscount = autoData.readF64();
        result.secondRefLinkDiscount = autoData.readF64();
        result.firstRefLinkRatio = autoData.readF64();
        result.secondRefLinkRatio = autoData.readF64();
        result.refProgramDiscount = autoData.readF64();
        result.refProgramRatio = autoData.readF64();
        result.reserved = autoData.readI64();
        result.mask = autoData.readI64();
        result.id = autoData.readU32();
        result.refClientId = autoData.readU32();
        result.refCounter = autoData.readU32();
        result.firstRefLinkId = autoData.readU32();
        result.secondRefLinkId = autoData.readU32();
        result.firstRefLinkExpiration = autoData.readU32();
        result.secondRefLinkExpiration = autoData.readU32();
        result.refProgramExpiration = autoData.readU32();
        result.spotTrades = autoData.readU32();
        result.perpTrades = autoData.readU32();
        result.lpTrades = autoData.readU32();
        result.points = autoData.readU32();
        result.slot = autoData.readU32();
        result.assetsCount = autoData.readU32();
        result.spotFilledOrders = autoData.readU32();
        result.perpFilledOrders = autoData.readU32();
        result.reservedValue1 = autoData.readI64();
        result.reservedValue2 = autoData.readI64();
        result.reservedValue3 = autoData.readI64();
        result.reservedValue4 = autoData.readI64();
        result.reservedValue5 = autoData.readI64();
        result.reservedValue6 = autoData.readI64();
        result.reservedValue7 = autoData.readI64();
        result.reservedValue8 = autoData.readI64();
        return result;
    }
}
exports.ClientPrimaryAccountHeaderModel = ClientPrimaryAccountHeaderModel;
ClientPrimaryAccountHeaderModel.LENGTH = 28 * 4 + 17 * 8 + 4 * 32; // 376 bytes
ClientPrimaryAccountHeaderModel.OFFSET_TAG = 0;
ClientPrimaryAccountHeaderModel.OFFSET_VERSION = 4;
ClientPrimaryAccountHeaderModel.OFFSET_WALLET_ADDRESS = 8;
ClientPrimaryAccountHeaderModel.OFFSET_LUT_ADDRESS = 40;
ClientPrimaryAccountHeaderModel.OFFSET_REF_ADDRESS = 72;
ClientPrimaryAccountHeaderModel.OFFSET_VM_WALLET_ADDRESS = 104;
ClientPrimaryAccountHeaderModel.OFFSET_VM_INSTR_0 = 136;
ClientPrimaryAccountHeaderModel.OFFSET_VM_INSTR_1 = 140;
ClientPrimaryAccountHeaderModel.OFFSET_VM_INSTR_2 = 144;
ClientPrimaryAccountHeaderModel.OFFSET_VM_INSTR_3 = 148;
ClientPrimaryAccountHeaderModel.OFFSET_VM_INSTR_4 = 152;
ClientPrimaryAccountHeaderModel.OFFSET_VM_INSTR_5 = 156;
ClientPrimaryAccountHeaderModel.OFFSET_VM_INSTR_6 = 160;
ClientPrimaryAccountHeaderModel.OFFSET_VM_INSTR_7 = 164;
ClientPrimaryAccountHeaderModel.OFFSET_VM_WITHDRAW_TOKEN_ID = 168;
ClientPrimaryAccountHeaderModel.OFFSET_VM_MASK = 172;
ClientPrimaryAccountHeaderModel.OFFSET_VM_WITHDRAW_AMOUNT = 176;
ClientPrimaryAccountHeaderModel.OFFSET_FIRST_REF_LINK_DISCOUNT = 184;
ClientPrimaryAccountHeaderModel.OFFSET_SECOND_REF_LINK_DISCOUNT = 192;
ClientPrimaryAccountHeaderModel.OFFSET_FIRST_REF_LINK_RATIO = 200;
ClientPrimaryAccountHeaderModel.OFFSET_SECOND_REF_LINK_RATIO = 208;
ClientPrimaryAccountHeaderModel.OFFSET_REF_PROGRAM_DISCOUNT = 216;
ClientPrimaryAccountHeaderModel.OFFSET_REF_PROGRAM_RATIO = 224;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED = 232;
ClientPrimaryAccountHeaderModel.OFFSET_MASK = 240;
ClientPrimaryAccountHeaderModel.OFFSET_ID = 248;
ClientPrimaryAccountHeaderModel.OFFSET_REF_CLIENT_ID = 252;
ClientPrimaryAccountHeaderModel.OFFSET_REF_COUNTER = 256;
ClientPrimaryAccountHeaderModel.OFFSET_FIRST_REF_LINK_ID = 260;
ClientPrimaryAccountHeaderModel.OFFSET_SECOND_REF_LINK_ID = 264;
ClientPrimaryAccountHeaderModel.OFFSET_FIRST_REF_LINK_EXPIRATION = 268;
ClientPrimaryAccountHeaderModel.OFFSET_SECOND_REF_LINK_EXPIRATION = 272;
ClientPrimaryAccountHeaderModel.OFFSET_REF_PROGRAM_EXPIRATION = 276;
ClientPrimaryAccountHeaderModel.OFFSET_SPOT_TRADES = 280;
ClientPrimaryAccountHeaderModel.OFFSET_PERP_TRADES = 284;
ClientPrimaryAccountHeaderModel.OFFSET_LP_TRADES = 288;
ClientPrimaryAccountHeaderModel.OFFSET_POINTS = 292;
ClientPrimaryAccountHeaderModel.OFFSET_SLOT = 296;
ClientPrimaryAccountHeaderModel.OFFSET_ASSETS_COUNT = 300;
ClientPrimaryAccountHeaderModel.OFFSET_SPOT_FILLED_ORDERS = 304;
ClientPrimaryAccountHeaderModel.OFFSET_PERP_FILLED_ORDERS = 308;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED_VALUE1 = 312;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED_VALUE2 = 320;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED_VALUE3 = 328;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED_VALUE4 = 336;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED_VALUE5 = 344;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED_VALUE6 = 352;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED_VALUE7 = 360;
ClientPrimaryAccountHeaderModel.OFFSET_RESERVED_VALUE8 = 368;
class OrderModel {
    static fromBuffer(data, offset) {
        const result = new OrderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.qty = autoData.readI64();
        result.sum = autoData.readI64();
        result.orderId = autoData.readI64();
        result.origClientId = autoData.readU32();
        result.clientId = autoData.readU32();
        result.line = autoData.readU32();
        result.prev = autoData.readU32();
        result.next = autoData.readU32();
        result.sref = autoData.readU32();
        result.link = autoData.readU32();
        result.clPrev = autoData.readU32();
        result.clNext = autoData.readU32();
        result.time = autoData.readU32();
        return result;
    }
}
exports.OrderModel = OrderModel;
OrderModel.LENGTH = 10 * 4 + 3 * 8; // 64 bytes
OrderModel.OFFSET_QTY = 0;
OrderModel.OFFSET_SUM = 8;
OrderModel.OFFSET_ORDER_ID = 16;
OrderModel.OFFSET_ORIG_CLIENT_ID = 24;
OrderModel.OFFSET_CLIENT_ID = 28;
OrderModel.OFFSET_LINE = 32;
OrderModel.OFFSET_PREV = 36;
OrderModel.OFFSET_NEXT = 40;
OrderModel.OFFSET_SREF = 44;
OrderModel.OFFSET_LINK = 48;
OrderModel.OFFSET_CL_PREV = 52;
OrderModel.OFFSET_CL_NEXT = 56;
OrderModel.OFFSET_TIME = 60;
class PxOrdersModel {
    static fromBuffer(data, offset) {
        const result = new PxOrdersModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.price = autoData.readI64();
        result.qty = autoData.readI64();
        result.next = autoData.readU32();
        result.prev = autoData.readU32();
        result.sref = autoData.readU32();
        result.link = autoData.readU32();
        result.begin = autoData.readU32();
        result.end = autoData.readU32();
        return result;
    }
}
exports.PxOrdersModel = PxOrdersModel;
PxOrdersModel.LENGTH = 6 * 4 + 2 * 8; // 40 bytes
PxOrdersModel.OFFSET_PRICE = 0;
PxOrdersModel.OFFSET_QTY = 8;
PxOrdersModel.OFFSET_NEXT = 16;
PxOrdersModel.OFFSET_PREV = 20;
PxOrdersModel.OFFSET_SREF = 24;
PxOrdersModel.OFFSET_LINK = 28;
PxOrdersModel.OFFSET_BEGIN = 32;
PxOrdersModel.OFFSET_END = 36;
class PrivateClientHeaderModel {
    static fromBuffer(data, offset) {
        const result = new PrivateClientHeaderModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.tag = autoData.readU32();
        result.version = autoData.readU32();
        return result;
    }
}
exports.PrivateClientHeaderModel = PrivateClientHeaderModel;
PrivateClientHeaderModel.LENGTH = 2 * 4; // 8 bytes
PrivateClientHeaderModel.OFFSET_TAG = 0;
PrivateClientHeaderModel.OFFSET_VERSION = 4;
class PrivateClientModel {
    static fromBuffer(data, offset) {
        const result = new PrivateClientModel();
        let autoData = new auto_data_1.AutoData(data, offset);
        result.creationTime = autoData.readU32();
        result.expirationTime = autoData.readU32();
        result.wallet = autoData.readAddress();
        return result;
    }
}
exports.PrivateClientModel = PrivateClientModel;
PrivateClientModel.LENGTH = 2 * 4 + 1 * 32; // 40 bytes
PrivateClientModel.OFFSET_CREATION_TIME = 0;
PrivateClientModel.OFFSET_EXPIRATION_TIME = 4;
PrivateClientModel.OFFSET_WALLET = 8;
