// @deriverse/kit/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MARKET_DEPTH = exports.PROGRAM_ID = exports.VERSION = exports.AccountType = exports.InstrMask = void 0;
const kit_1 = require("@solana/kit");
var InstrMask;
(function (InstrMask) {
    InstrMask[InstrMask["DRV"] = 268435456] = "DRV";
    InstrMask[InstrMask["READY_TO_DRV_UPGRADE"] = 536870912] = "READY_TO_DRV_UPGRADE";
    InstrMask[InstrMask["PERP"] = 1073741824] = "PERP";
    InstrMask[InstrMask["ORACLE"] = 2147483648] = "ORACLE";
    InstrMask[InstrMask["READY_TO_PERP_UPGRADE"] = 16777216] = "READY_TO_PERP_UPGRADE";
})(InstrMask || (exports.InstrMask = InstrMask = {}));
var AccountType;
(function (AccountType) {
    AccountType[AccountType["CLIENT_COMMUNITY"] = 35] = "CLIENT_COMMUNITY";
    AccountType[AccountType["CLIENT_PRIMARY"] = 31] = "CLIENT_PRIMARY";
    AccountType[AccountType["COMMUNITY"] = 34] = "COMMUNITY";
    AccountType[AccountType["HOLDER"] = 1] = "HOLDER";
    AccountType[AccountType["ROOT"] = 2] = "ROOT";
    AccountType[AccountType["INSTR"] = 7] = "INSTR";
    AccountType[AccountType["SPOT_15M_CANDLES"] = 20] = "SPOT_15M_CANDLES";
    AccountType[AccountType["SPOT_1M_CANDLES"] = 19] = "SPOT_1M_CANDLES";
    AccountType[AccountType["SPOT_ASK_ORDERS"] = 17] = "SPOT_ASK_ORDERS";
    AccountType[AccountType["SPOT_ASKS_TREE"] = 15] = "SPOT_ASKS_TREE";
    AccountType[AccountType["SPOT_BID_ORDERS"] = 16] = "SPOT_BID_ORDERS";
    AccountType[AccountType["SPOT_BIDS_TREE"] = 14] = "SPOT_BIDS_TREE";
    AccountType[AccountType["SPOT_CLIENT_INFOS"] = 12] = "SPOT_CLIENT_INFOS";
    AccountType[AccountType["SPOT_CLIENT_INFOS2"] = 13] = "SPOT_CLIENT_INFOS2";
    AccountType[AccountType["SPOT_DAY_CANDLES"] = 21] = "SPOT_DAY_CANDLES";
    AccountType[AccountType["SPOT_LINES"] = 18] = "SPOT_LINES";
    AccountType[AccountType["SPOT_MAPS"] = 10] = "SPOT_MAPS";
    AccountType[AccountType["TOKEN"] = 4] = "TOKEN";
    AccountType[AccountType["PERP_ASK_ORDERS"] = 36] = "PERP_ASK_ORDERS";
    AccountType[AccountType["PERP_ASKS_TREE"] = 37] = "PERP_ASKS_TREE";
    AccountType[AccountType["PERP_BID_ORDERS"] = 38] = "PERP_BID_ORDERS";
    AccountType[AccountType["PERP_BIDS_TREE"] = 39] = "PERP_BIDS_TREE";
    AccountType[AccountType["PERP_CLIENT_INFOS"] = 41] = "PERP_CLIENT_INFOS";
    AccountType[AccountType["PERP_CLIENT_INFOS2"] = 42] = "PERP_CLIENT_INFOS2";
    AccountType[AccountType["PERP_CLIENT_INFOS3"] = 43] = "PERP_CLIENT_INFOS3";
    AccountType[AccountType["PERP_CLIENT_INFOS4"] = 44] = "PERP_CLIENT_INFOS4";
    AccountType[AccountType["PERP_CLIENT_INFOS5"] = 45] = "PERP_CLIENT_INFOS5";
    AccountType[AccountType["PERP_LINES"] = 46] = "PERP_LINES";
    AccountType[AccountType["PERP_MAPS"] = 47] = "PERP_MAPS";
    AccountType[AccountType["PERP_LONG_PX_TREE"] = 48] = "PERP_LONG_PX_TREE";
    AccountType[AccountType["PERP_SHORT_PX_TREE"] = 49] = "PERP_SHORT_PX_TREE";
    AccountType[AccountType["PERP_REBALANCE_TIME_TREE"] = 50] = "PERP_REBALANCE_TIME_TREE";
    AccountType[AccountType["PRIVATE_CLIENTS"] = 51] = "PRIVATE_CLIENTS";
})(AccountType || (exports.AccountType = AccountType = {}));
exports.VERSION = 1;
exports.PROGRAM_ID = (0, kit_1.address)("DRVSpZ2YUYYKgZP8XtLhAGtT1zYSCKzeHfb4DgRnrgqD");
exports.MARKET_DEPTH = 20;
